// @redux-devtools/ui/lib/cjs/Tabs/data.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.tabs = exports.simple10Tabs = void 0;

var _react = _interopRequireDefault(require("react"));

/* eslint-disable react/prop-types */
var Component = function Component(_ref) {
  var selected = _ref.selected;
  return /*#__PURE__*/_react.default.createElement("div", {
    style: {
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center',
      width: '100%',
      height: '100%',
      fontSize: '22px'
    }
  }, "Selected ", selected);
};
/* eslint-enable react/prop-types */


var selector = function selector(tab) {
  return {
    selected: tab.name
  };
};

var tabs = [{
  name: 'Tab1',
  component: Component,
  selector: selector
}, {
  name: 'Tab2',
  component: Component,
  selector: selector
}, {
  name: 'Tab3',
  component: Component,
  selector: selector
}];
exports.tabs = tabs;
var simple10Tabs = [];
exports.simple10Tabs = simple10Tabs;

for (var i = 1; i <= 10; i++) {
  simple10Tabs.push({
    name: "Tab".concat(i),
    value: "".concat(i)
  });
}
