// @redux-devtools/ui/lib/cjs/Tabs/Tabs.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.WithContent = exports.Default = void 0;

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _ = _interopRequireDefault(require("./"));

var _data = require("./data");

var _templateObject;

var Container = _styledComponents.default.div(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2.default)(["\n  display: flex;\n  height: 100%;\n  width: 100%;\n  justify-content: center;\n  align-items: center;\n"])));

var _default = {
  title: 'Tabs',
  component: _.default
};
exports.default = _default;

var Template = function Template(args) {
  return /*#__PURE__*/_react.default.createElement(Container, null, /*#__PURE__*/_react.default.createElement(_.default, args));
};

var Default = Template.bind({});
exports.Default = Default;
Default.args = {
  tabs: _data.simple10Tabs,
  selected: '2',
  main: true,
  collapsible: true,
  position: 'left'
};
Default.argTypes = {
  tabs: {
    control: {
      disable: true
    }
  },
  onClick: {
    control: {
      disable: true
    }
  }
};
var WithContent = Template.bind({});
exports.WithContent = WithContent;
WithContent.args = {
  tabs: _data.tabs,
  selected: 'Tab2',
  main: false,
  collapsible: false,
  position: 'left'
};
WithContent.argTypes = {
  tabs: {
    control: {
      disable: true
    }
  },
  onClick: {
    control: {
      disable: true
    }
  }
};
