// @redux-devtools/ui/lib/cjs/Tabs/Tabs.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _TabsHeader = _interopRequireDefault(require("./TabsHeader"));

var _common = require("./styles/common");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var Tabs = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Tabs, _Component);

  var _super = _createSuper(Tabs);

  function Tabs() {
    var _this;

    (0, _classCallCheck2.default)(this, Tabs);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onMouseUp", function (e) {
      e.currentTarget.blur();
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onClick", function (e) {
      _this.props.onClick(e.currentTarget.value);
    });
    return _this;
  }

  (0, _createClass2.default)(Tabs, [{
    key: "renderTabs",
    value: function renderTabs() {
      var _this2 = this;

      var tabs = this.props.tabs;
      var selected = this.props.selected;
      return tabs.map(function (tab, i) {
        var isSelected;
        var value = typeof tab.value !== 'undefined' ? tab.value : tab.name;

        if (value === selected) {
          isSelected = true;

          if (tab.component) {
            _this2.SelectedComponent = tab.component;
            if (tab.selector) _this2.selector = function () {
              return tab.selector(tab);
            };
          }
        }

        return /*#__PURE__*/_react.default.createElement("button", {
          key: i,
          value: value,
          "data-selected": isSelected,
          onMouseUp: _this2.onMouseUp,
          onClick: _this2.onClick
        }, tab.name);
      });
    }
  }, {
    key: "render",
    value: function render() {
      var tabsHeader = /*#__PURE__*/_react.default.createElement(_TabsHeader.default, {
        tabs: this.renderTabs(),
        items: this.props.tabs,
        main: this.props.main,
        collapsible: this.props.collapsible,
        onClick: this.props.onClick,
        selected: this.props.selected,
        position: this.props.position
      });

      if (!this.SelectedComponent) {
        return /*#__PURE__*/_react.default.createElement(_common.TabsContainer, {
          position: this.props.position
        }, tabsHeader);
      }

      return /*#__PURE__*/_react.default.createElement(_common.TabsContainer, {
        position: this.props.position
      }, tabsHeader, /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(this.SelectedComponent, this.selector && this.selector())));
    }
  }]);
  return Tabs;
}(_react.Component);

exports.default = Tabs;
(0, _defineProperty2.default)(Tabs, "propTypes", {
  tabs: _propTypes.default.array.isRequired,
  selected: _propTypes.default.string,
  main: _propTypes.default.bool,
  onClick: _propTypes.default.func.isRequired,
  collapsible: _propTypes.default.bool,
  position: _propTypes.default.oneOf(['left', 'right', 'center'])
});
(0, _defineProperty2.default)(Tabs, "defaultProps", {
  position: 'left'
});
