// @redux-devtools/ui/lib/cjs/Tabs/TabsHeader.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _simpleElementResizeDetector = _interopRequireDefault(require("simple-element-resize-detector"));

var _fa = require("react-icons/fa");

var _ContextMenu = _interopRequireDefault(require("../ContextMenu"));

var _createStyledComponent = _interopRequireDefault(require("../utils/createStyledComponent"));

var styles = _interopRequireWildcard(require("./styles"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var TabsWrapper = (0, _createStyledComponent.default)(styles);

var TabsHeader = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(TabsHeader, _Component);

  var _super = _createSuper(TabsHeader);

  function TabsHeader() {
    var _this;

    (0, _classCallCheck2.default)(this, TabsHeader);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "state", {
      visibleTabs: _this.props.tabs.slice(),
      hiddenTabs: [],
      subMenuOpened: false,
      contextMenu: undefined
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "iconWidth", 0);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "hiddenTabsWidth", []);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "collapse", function () {
      if (_this.state.subMenuOpened) _this.hideSubmenu();
      var _this$props = _this.props,
          selected = _this$props.selected,
          tabs = _this$props.tabs;
      var tabsWrapperRef = _this.tabsWrapperRef;
      var tabsRef = _this.tabsRef;
      var tabButtons = _this.tabsRef.children;
      var visibleTabs = _this.state.visibleTabs;
      var hiddenTabs = _this.state.hiddenTabs;
      var tabsWrapperRight = tabsWrapperRef.getBoundingClientRect().right;
      if (!tabsWrapperRight) return; // tabs are hidden

      var tabsRefRight = tabsRef.getBoundingClientRect().right;
      var i = visibleTabs.length - 1;
      var hiddenTab;

      if (tabsRefRight >= tabsWrapperRight - _this.iconWidth) {
        if (_this.props.position === 'right' && hiddenTabs.length > 0 && tabsRef.getBoundingClientRect().width + _this.hiddenTabsWidth[0] < tabsWrapperRef.getBoundingClientRect().width) {
          while (i < tabs.length - 1 && tabsRef.getBoundingClientRect().width + _this.hiddenTabsWidth[0] < tabsWrapperRef.getBoundingClientRect().width) {
            hiddenTab = hiddenTabs.shift();
            visibleTabs.splice(Number(hiddenTab.key), 0, hiddenTab);
            i++;
          }
        } else {
          while (i > 0 && tabButtons[i] && tabButtons[i].getBoundingClientRect().right >= tabsWrapperRight - _this.iconWidth) {
            if (tabButtons[i].value !== selected) {
              hiddenTabs.unshift.apply(hiddenTabs, (0, _toConsumableArray2.default)(visibleTabs.splice(i, 1)));

              _this.hiddenTabsWidth.unshift(tabButtons[i].getBoundingClientRect().width);
            } else {
              tabsWrapperRight -= tabButtons[i].getBoundingClientRect().width;
            }

            i--;
          }
        }
      } else {
        while (i < tabs.length - 1 && tabButtons[i] && tabButtons[i].getBoundingClientRect().right + _this.hiddenTabsWidth[0] < tabsWrapperRight - _this.iconWidth) {
          hiddenTab = hiddenTabs.shift();
          visibleTabs.splice(Number(hiddenTab.key), 0, hiddenTab);

          _this.hiddenTabsWidth.shift();

          i++;
        }
      }

      _this.setState({
        visibleTabs: visibleTabs,
        hiddenTabs: hiddenTabs
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "hideSubmenu", function () {
      _this.setState({
        subMenuOpened: false,
        contextMenu: undefined
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "getTabsWrapperRef", function (node) {
      _this.tabsWrapperRef = node;
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "getTabsRef", function (node) {
      _this.tabsRef = node;
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "expandMenu", function (e) {
      var rect = e.currentTarget.children[0].getBoundingClientRect();

      _this.setState({
        contextMenu: {
          top: rect.top + 10,
          left: rect.left
        },
        subMenuOpened: true
      });
    });
    return _this;
  }

  (0, _createClass2.default)(TabsHeader, [{
    key: "UNSAFE_componentWillReceiveProps",
    value: function UNSAFE_componentWillReceiveProps(nextProps) {
      if (nextProps.tabs !== this.props.tabs || nextProps.selected !== this.props.selected || nextProps.collapsible !== this.props.collapsible) {
        this.setState({
          hiddenTabs: [],
          visibleTabs: nextProps.tabs.slice()
        });
      }
    }
  }, {
    key: "componentDidMount",
    value: function componentDidMount() {
      if (this.props.collapsible) {
        this.collapse();
        this.enableResizeEvents();
      }
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      var collapsible = this.props.collapsible;

      if (!collapsible) {
        if (prevProps.collapsible !== collapsible) this.disableResizeEvents();
        return;
      }

      var shouldCollapse = false;

      if (this.iconWidth === 0) {
        var tabButtons = this.tabsRef.children;

        if (this.tabsRef.children[tabButtons.length - 1].value === 'expandIcon') {
          this.iconWidth = tabButtons[tabButtons.length - 1].getBoundingClientRect().width;
          shouldCollapse = true;
        }
      } else if (this.state.hiddenTabs.length === 0) {
        this.iconWidth = 0;
      }

      if (prevProps.collapsible !== collapsible) {
        this.enableResizeEvents();
        shouldCollapse = true;
      }

      if (shouldCollapse || this.props.selected !== prevProps.selected) {
        this.collapse();
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      if (this.props.collapsible) {
        this.disableResizeEvents();
      }
    }
  }, {
    key: "enableResizeEvents",
    value: function enableResizeEvents() {
      this.resizeDetector = (0, _simpleElementResizeDetector.default)(this.tabsWrapperRef, this.collapse);
      window.addEventListener('mousedown', this.hideSubmenu);
    }
  }, {
    key: "disableResizeEvents",
    value: function disableResizeEvents() {
      this.resizeDetector.remove();
      window.removeEventListener('mousedown', this.hideSubmenu);
    }
  }, {
    key: "render",
    value: function render() {
      var _this$state = this.state,
          visibleTabs = _this$state.visibleTabs,
          hiddenTabs = _this$state.hiddenTabs,
          contextMenu = _this$state.contextMenu;
      return /*#__PURE__*/_react.default.createElement(TabsWrapper, {
        ref: this.getTabsWrapperRef,
        main: this.props.main,
        position: this.props.position
      }, /*#__PURE__*/_react.default.createElement("div", {
        ref: this.getTabsRef
      }, visibleTabs, this.props.collapsible && visibleTabs.length < this.props.items.length && /*#__PURE__*/_react.default.createElement("button", {
        onClick: this.expandMenu,
        value: "expandIcon"
      }, /*#__PURE__*/_react.default.createElement(_fa.FaAngleDoubleRight, null))), this.props.collapsible && contextMenu && /*#__PURE__*/_react.default.createElement(_ContextMenu.default, {
        items: hiddenTabs,
        onClick: this.props.onClick,
        x: contextMenu.left,
        y: contextMenu.top,
        visible: this.state.subMenuOpened
      }));
    }
  }]);
  return TabsHeader;
}(_react.Component);

exports.default = TabsHeader;
(0, _defineProperty2.default)(TabsHeader, "propTypes", {
  tabs: _propTypes.default.array.isRequired,
  items: _propTypes.default.array.isRequired,
  main: _propTypes.default.bool,
  onClick: _propTypes.default.func,
  position: _propTypes.default.string,
  collapsible: _propTypes.default.bool,
  selected: _propTypes.default.string
});
