// @redux-devtools/ui/lib/esm/Button/Button.stories.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject;

import React from 'react';
import styled from 'styled-components';
import { MdFiberManualRecord } from 'react-icons/md';
import Button from './';
export default {
  title: 'Button',
  component: Button
};
var Container = styled.div(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  display: flex;\n  height: 100%;\n  width: 100%;\n  justify-content: center;\n  align-items: center;\n"])));

var Template = function Template(args) {
  return /*#__PURE__*/React.createElement(Container, null, /*#__PURE__*/React.createElement(Button, args));
};

export var Default = Template.bind({});
Default.args = {
  title: 'Hello Tooltip! \\a And from new line hello!',
  tooltipPosition: 'top',
  primary: true,
  size: 'normal',
  disabled: false,
  children: 'Hello Button'
};
Default.argTypes = {
  onClick: {
    control: {
      disable: true
    }
  },
  type: {
    control: {
      disable: true
    }
  },
  mark: {
    control: {
      disable: true
    }
  },
  theme: {
    control: {
      disable: true
    }
  }
};
export var Mark = Template.bind({});
Mark.args = {
  mark: 'base08',
  title: 'Hello Tooltip',
  tooltipPosition: 'top',
  size: 'normal',
  disabled: false,
  children: /*#__PURE__*/React.createElement(MdFiberManualRecord, null)
};
Mark.argTypes = {
  children: {
    control: {
      disable: true
    }
  },
  onClick: {
    control: {
      disable: true
    }
  },
  type: {
    control: {
      disable: true
    }
  },
  primary: {
    control: {
      disable: true
    }
  },
  theme: {
    control: {
      disable: true
    }
  }
};
