// @redux-devtools/log-monitor/lib/cjs/LogMonitorEntry.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactJsonTree = require("react-json-tree");

var _LogMonitorEntryAction = _interopRequireDefault(require("./LogMonitorEntryAction"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var styles = {
  entry: {
    display: 'block',
    WebkitUserSelect: 'none'
  },
  root: {
    marginLeft: 0
  }
};

var getDeepItem = function getDeepItem(data, path) {
  return path.reduce(function (obj, key) {
    return obj && obj[key];
  }, data);
};

var dataIsEqual = function dataIsEqual(data, previousData, keyPath) {
  var path = (0, _toConsumableArray2.default)(keyPath).reverse().slice(1);
  return getDeepItem(data, path) === getDeepItem(previousData, path);
};

var LogMonitorEntry = /*#__PURE__*/function (_PureComponent) {
  (0, _inherits2.default)(LogMonitorEntry, _PureComponent);

  var _super = _createSuper(LogMonitorEntry);

  function LogMonitorEntry() {
    var _this;

    (0, _classCallCheck2.default)(this, LogMonitorEntry);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleActionClick", function (e) {
      var _this$props = _this.props,
          actionId = _this$props.actionId,
          onActionClick = _this$props.onActionClick,
          onActionShiftClick = _this$props.onActionShiftClick;

      if (actionId > 0) {
        if (e.shiftKey) {
          onActionShiftClick(actionId);
        } else {
          onActionClick(actionId);
        }
      }
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "shouldExpandNode", function (keyPath, data, level) {
      return _this.props.expandStateRoot && level === 0;
    });
    return _this;
  }

  (0, _createClass2.default)(LogMonitorEntry, [{
    key: "printState",
    value: function printState(state, error) {
      var _this2 = this;

      var errorText = error;

      if (!errorText) {
        try {
          var data = this.props.select(state);
          var theme;

          if (this.props.markStateDiff) {
            var previousData = typeof this.props.previousState !== 'undefined' ? this.props.select(this.props.previousState) : undefined;

            var getValueStyle = function getValueStyle(_ref, nodeType, keyPath) {
              var style = _ref.style;
              return {
                style: _objectSpread(_objectSpread({}, style), {}, {
                  backgroundColor: dataIsEqual(data, previousData, keyPath) ? 'transparent' : _this2.props.theme.base01
                })
              };
            };

            var getNestedNodeStyle = function getNestedNodeStyle(_ref2, keyPath) {
              var style = _ref2.style;
              return {
                style: _objectSpread(_objectSpread({}, style), keyPath.length > 1 ? {} : styles.root)
              };
            };

            theme = {
              extend: this.props.theme,
              value: getValueStyle,
              nestedNode: getNestedNodeStyle
            };
          } else {
            theme = this.props.theme;
          }

          return /*#__PURE__*/_react.default.createElement(_reactJsonTree.JSONTree, {
            theme: theme,
            data: data,
            invertTheme: false,
            keyPath: ['state'],
            shouldExpandNode: this.shouldExpandNode
          });
        } catch (err) {
          errorText = 'Error selecting state.';
        }
      }

      return /*#__PURE__*/_react.default.createElement("div", {
        style: {
          color: this.props.theme.base08,
          paddingTop: 20,
          paddingLeft: 30,
          paddingRight: 30,
          paddingBottom: 35
        }
      }, errorText);
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          actionId = _this$props2.actionId,
          error = _this$props2.error,
          action = _this$props2.action,
          state = _this$props2.state,
          collapsed = _this$props2.collapsed,
          selected = _this$props2.selected,
          inFuture = _this$props2.inFuture;
      var styleEntry = {
        opacity: collapsed ? 0.5 : 1,
        cursor: actionId > 0 ? 'pointer' : 'default'
      };
      return /*#__PURE__*/_react.default.createElement("div", {
        style: {
          opacity: selected ? 0.4 : inFuture ? 0.6 : 1,
          // eslint-disable-line no-nested-ternary
          textDecoration: collapsed ? 'line-through' : 'none',
          color: this.props.theme.base06
        }
      }, /*#__PURE__*/_react.default.createElement(_LogMonitorEntryAction.default, {
        theme: this.props.theme,
        collapsed: collapsed,
        action: action,
        expandActionRoot: this.props.expandActionRoot,
        onClick: this.handleActionClick,
        style: _objectSpread(_objectSpread({}, styles.entry), styleEntry)
      }), !collapsed && /*#__PURE__*/_react.default.createElement("div", {
        style: {
          paddingLeft: 16
        }
      }, this.printState(state, error)));
    }
  }]);
  return LogMonitorEntry;
}(_react.PureComponent);

exports.default = LogMonitorEntry;
(0, _defineProperty2.default)(LogMonitorEntry, "propTypes", {
  state: _propTypes.default.object.isRequired,
  action: _propTypes.default.object.isRequired,
  actionId: _propTypes.default.number.isRequired,
  select: _propTypes.default.func.isRequired,
  inFuture: _propTypes.default.bool,
  error: _propTypes.default.string,
  onActionClick: _propTypes.default.func.isRequired,
  onActionShiftClick: _propTypes.default.func.isRequired,
  collapsed: _propTypes.default.bool,
  selected: _propTypes.default.bool,
  expandActionRoot: _propTypes.default.bool,
  expandStateRoot: _propTypes.default.bool,
  previousState: _propTypes.default.object
});
