// @redux-devtools/ui/lib/esm/Dialog/Dialog.stories.js
import React from 'react';
import Dialog from './';
import { schema, uiSchema, formData } from '../Form/schema';
export default {
  title: 'Dialog',
  component: Dialog
};

var Template = function Template(args) {
  return /*#__PURE__*/React.createElement(Dialog, args);
};

export var Default = Template.bind({});
Default.args = {
  title: 'Dialog Title',
  submitText: 'Submit!',
  open: true,
  noHeader: false,
  noFooter: false,
  modal: false,
  fullWidth: false,
  children: 'Hello Dialog!'
};
Default.argTypes = {
  actions: {
    control: {
      disable: true
    }
  },
  onDismiss: {
    control: {
      disable: true
    }
  },
  onSubmit: {
    control: {
      disable: true
    }
  },
  theme: {
    control: {
      disable: true
    }
  }
};
export var WithForm = Template.bind({});
WithForm.args = {
  open: true,
  noHeader: false,
  noFooter: false,
  fullWidth: false,
  submitText: 'Submit!',
  formData: formData,
  schema: schema,
  uiSchema: uiSchema
};
WithForm.argTypes = {
  title: {
    control: {
      disable: true
    }
  },
  children: {
    control: {
      disable: true
    }
  },
  actions: {
    control: {
      disable: true
    }
  },
  modal: {
    control: {
      disable: true
    }
  },
  onDismiss: {
    control: {
      disable: true
    }
  },
  onSubmit: {
    control: {
      disable: true
    }
  },
  theme: {
    control: {
      disable: true
    }
  }
};
