// @redux-devtools/app/lib/cjs/constants/actionTypes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UPDATE_STATE = exports.UPDATE_REPORTS = exports.UPDATE_MONITOR_STATE = exports.TOGGLE_SYNC = exports.TOGGLE_SLIDER = exports.TOGGLE_PERSIST = exports.TOGGLE_DISPATCHER = exports.SHOW_NOTIFICATION = exports.SET_STATE = exports.SET_PERSIST = exports.SELECT_MONITOR = exports.SELECT_INSTANCE = exports.REMOVE_INSTANCE = exports.MONITOR_ACTION = exports.LIFTED_ACTION = exports.GET_REPORT_SUCCESS = exports.GET_REPORT_REQUEST = exports.GET_REPORT_ERROR = exports.EXPORT = exports.ERROR = exports.CLEAR_NOTIFICATION = exports.CHANGE_THEME = exports.CHANGE_SECTION = void 0;
var CHANGE_SECTION = 'main/CHANGE_SECTION';
exports.CHANGE_SECTION = CHANGE_SECTION;
var CHANGE_THEME = 'main/CHANGE_THEME';
exports.CHANGE_THEME = CHANGE_THEME;
var UPDATE_STATE = 'devTools/UPDATE_STATE';
exports.UPDATE_STATE = UPDATE_STATE;
var SET_STATE = 'devTools/SET_STATE';
exports.SET_STATE = SET_STATE;
var SELECT_INSTANCE = 'devTools/SELECT_INSTANCE';
exports.SELECT_INSTANCE = SELECT_INSTANCE;
var REMOVE_INSTANCE = 'devTools/REMOVE_INSTANCE';
exports.REMOVE_INSTANCE = REMOVE_INSTANCE;
var LIFTED_ACTION = 'devTools/LIFTED_ACTION';
exports.LIFTED_ACTION = LIFTED_ACTION;
var MONITOR_ACTION = 'devTools/MONITOR_ACTION';
exports.MONITOR_ACTION = MONITOR_ACTION;
var TOGGLE_SYNC = 'devTools/TOGGLE_SYNC';
exports.TOGGLE_SYNC = TOGGLE_SYNC;
var TOGGLE_PERSIST = 'devTools/TOGGLE_PERSIST';
exports.TOGGLE_PERSIST = TOGGLE_PERSIST;
var SET_PERSIST = 'devTools/SET_PERSIST';
exports.SET_PERSIST = SET_PERSIST;
var SELECT_MONITOR = 'devTools/SELECT_MONITOR';
exports.SELECT_MONITOR = SELECT_MONITOR;
var UPDATE_MONITOR_STATE = 'devTools/UPDATE_MONITOR_STATE';
exports.UPDATE_MONITOR_STATE = UPDATE_MONITOR_STATE;
var TOGGLE_SLIDER = 'devTools/TOGGLE_SLIDER';
exports.TOGGLE_SLIDER = TOGGLE_SLIDER;
var TOGGLE_DISPATCHER = 'devTools/TOGGLE_DISPATCHER';
exports.TOGGLE_DISPATCHER = TOGGLE_DISPATCHER;
var EXPORT = 'devTools/EXPORT';
exports.EXPORT = EXPORT;
var SHOW_NOTIFICATION = 'devTools/SHOW_NOTIFICATION';
exports.SHOW_NOTIFICATION = SHOW_NOTIFICATION;
var CLEAR_NOTIFICATION = 'devTools/CLEAR_NOTIFICATION';
exports.CLEAR_NOTIFICATION = CLEAR_NOTIFICATION;
var UPDATE_REPORTS = 'reports/UPDATE';
exports.UPDATE_REPORTS = UPDATE_REPORTS;
var GET_REPORT_REQUEST = 'reports/GET_REPORT_REQUEST';
exports.GET_REPORT_REQUEST = GET_REPORT_REQUEST;
var GET_REPORT_ERROR = 'reports/GET_REPORT_ERROR';
exports.GET_REPORT_ERROR = GET_REPORT_ERROR;
var GET_REPORT_SUCCESS = 'reports/GET_REPORT_SUCCESS';
exports.GET_REPORT_SUCCESS = GET_REPORT_SUCCESS;
var ERROR = 'ERROR';
exports.ERROR = ERROR;
