// @redux-devtools/app/lib/cjs/constants/socketActionTypes.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UNSUBSCRIBE = exports.UNAUTHENTICATED = exports.SUBSCRIBE_SUCCESS = exports.SUBSCRIBE_REQUEST = exports.SUBSCRIBE_ERROR = exports.RECONNECT = exports.PENDING = exports.OPEN = exports.EMIT = exports.DISCONNECTED = exports.DEAUTHENTICATE = exports.CONNECT_SUCCESS = exports.CONNECT_REQUEST = exports.CONNECT_ERROR = exports.CONNECTING = exports.CLOSED = exports.AUTH_SUCCESS = exports.AUTH_REQUEST = exports.AUTH_ERROR = exports.AUTHENTICATED = void 0;

var _socketclusterClient = _interopRequireDefault(require("socketcluster-client"));

var _ref = _socketclusterClient.default.SCClientSocket,
    CLOSED = _ref.CLOSED,
    CONNECTING = _ref.CONNECTING,
    OPEN = _ref.OPEN,
    AUTHENTICATED = _ref.AUTHENTICATED,
    PENDING = _ref.PENDING,
    UNAUTHENTICATED = _ref.UNAUTHENTICATED;
exports.UNAUTHENTICATED = UNAUTHENTICATED;
exports.PENDING = PENDING;
exports.AUTHENTICATED = AUTHENTICATED;
exports.OPEN = OPEN;
exports.CONNECTING = CONNECTING;
exports.CLOSED = CLOSED;
var CONNECT_REQUEST = 'socket/CONNECT_REQUEST';
exports.CONNECT_REQUEST = CONNECT_REQUEST;
var CONNECT_SUCCESS = 'socket/CONNECT_SUCCESS';
exports.CONNECT_SUCCESS = CONNECT_SUCCESS;
var CONNECT_ERROR = 'socket/CONNECT_ERROR';
exports.CONNECT_ERROR = CONNECT_ERROR;
var RECONNECT = 'socket/RECONNECT';
exports.RECONNECT = RECONNECT;
var AUTH_REQUEST = 'socket/AUTH_REQUEST';
exports.AUTH_REQUEST = AUTH_REQUEST;
var AUTH_SUCCESS = 'socket/AUTH_SUCCESS';
exports.AUTH_SUCCESS = AUTH_SUCCESS;
var AUTH_ERROR = 'socket/AUTH_ERROR';
exports.AUTH_ERROR = AUTH_ERROR;
var DISCONNECTED = 'socket/DISCONNECTED';
exports.DISCONNECTED = DISCONNECTED;
var DEAUTHENTICATE = 'socket/DEAUTHENTICATE';
exports.DEAUTHENTICATE = DEAUTHENTICATE;
var SUBSCRIBE_REQUEST = 'socket/SUBSCRIBE_REQUEST';
exports.SUBSCRIBE_REQUEST = SUBSCRIBE_REQUEST;
var SUBSCRIBE_SUCCESS = 'socket/SUBSCRIBE_SUCCESS';
exports.SUBSCRIBE_SUCCESS = SUBSCRIBE_SUCCESS;
var SUBSCRIBE_ERROR = 'socket/SUBSCRIBE_ERROR';
exports.SUBSCRIBE_ERROR = SUBSCRIBE_ERROR;
var UNSUBSCRIBE = 'socket/UNSUBSCRIBE';
exports.UNSUBSCRIBE = UNSUBSCRIBE;
var EMIT = 'socket/EMIT';
exports.EMIT = EMIT;
