// @redux-devtools/ui/lib/esm/Editor/Editor.stories.js
import React from 'react';
import Editor from './';
import { default as WithTabsComponent } from './WithTabs';
var value = "\nvar themes = [];\n\nfunction getThemes() {\n  return themes;\n}\n";
export default {
  title: 'Editor',
  component: Editor
};

var Template = function Template(args) {
  return /*#__PURE__*/React.createElement(Editor, args);
};

export var Default = Template.bind({});
Default.args = {
  value: value,
  lineNumbers: true,
  lineWrapping: false,
  foldGutter: true,
  readOnly: false,
  autofocus: true
};
Default.argTypes = {
  autofocus: {
    control: {
      disable: true
    }
  },
  mode: {
    control: {
      disable: true
    }
  },
  theme: {
    control: {
      disable: true
    }
  },
  onChange: {
    control: {
      disable: true
    }
  }
};

var WithTabsTemplate = function WithTabsTemplate(args) {
  return /*#__PURE__*/React.createElement(WithTabsComponent, args);
};

export var WithTabs = WithTabsTemplate.bind({});
WithTabs.args = {
  lineNumbers: true,
  position: 'left'
};
WithTabs.argTypes = {
  value: {
    control: {
      disable: true
    }
  },
  mode: {
    control: {
      disable: true
    }
  },
  lineWrapping: {
    control: {
      disable: true
    }
  },
  readOnly: {
    control: {
      disable: true
    }
  },
  theme: {
    control: {
      disable: true
    }
  },
  foldGutter: {
    control: {
      disable: true
    }
  },
  autofocus: {
    control: {
      disable: true
    }
  },
  onChange: {
    control: {
      disable: true
    }
  }
};
