// @redux-devtools/app/lib/esm/constants/actionTypes.js
export var CHANGE_SECTION = 'main/CHANGE_SECTION';
export var CHANGE_THEME = 'main/CHANGE_THEME';
export var UPDATE_STATE = 'devTools/UPDATE_STATE';
export var SET_STATE = 'devTools/SET_STATE';
export var SELECT_INSTANCE = 'devTools/SELECT_INSTANCE';
export var REMOVE_INSTANCE = 'devTools/REMOVE_INSTANCE';
export var LIFTED_ACTION = 'devTools/LIFTED_ACTION';
export var MONITOR_ACTION = 'devTools/MONITOR_ACTION';
export var TOGGLE_SYNC = 'devTools/TOGGLE_SYNC';
export var TOGGLE_PERSIST = 'devTools/TOGGLE_PERSIST';
export var SET_PERSIST = 'devTools/SET_PERSIST';
export var SELECT_MONITOR = 'devTools/SELECT_MONITOR';
export var UPDATE_MONITOR_STATE = 'devTools/UPDATE_MONITOR_STATE';
export var TOGGLE_SLIDER = 'devTools/TOGGLE_SLIDER';
export var TOGGLE_DISPATCHER = 'devTools/TOGGLE_DISPATCHER';
export var EXPORT = 'devTools/EXPORT';
export var SHOW_NOTIFICATION = 'devTools/SHOW_NOTIFICATION';
export var CLEAR_NOTIFICATION = 'devTools/CLEAR_NOTIFICATION';
export var UPDATE_REPORTS = 'reports/UPDATE';
export var GET_REPORT_REQUEST = 'reports/GET_REPORT_REQUEST';
export var GET_REPORT_ERROR = 'reports/GET_REPORT_ERROR';
export var GET_REPORT_SUCCESS = 'reports/GET_REPORT_SUCCESS';
export var ERROR = 'ERROR';
