// @redux-devtools/app/lib/esm/constants/socketActionTypes.js
import socketCluster from 'socketcluster-client';
var _ref = socketCluster.SCClientSocket,
    CLOSED = _ref.CLOSED,
    CONNECTING = _ref.CONNECTING,
    OPEN = _ref.OPEN,
    AUTHENTICATED = _ref.AUTHENTICATED,
    PENDING = _ref.PENDING,
    UNAUTHENTICATED = _ref.UNAUTHENTICATED;
export { CLOSED, CONNECTING, OPEN, AUTHENTICATED, PENDING, UNAUTHENTICATED };
export var CONNECT_REQUEST = 'socket/CONNECT_REQUEST';
export var CONNECT_SUCCESS = 'socket/CONNECT_SUCCESS';
export var CONNECT_ERROR = 'socket/CONNECT_ERROR';
export var RECONNECT = 'socket/RECONNECT';
export var AUTH_REQUEST = 'socket/AUTH_REQUEST';
export var AUTH_SUCCESS = 'socket/AUTH_SUCCESS';
export var AUTH_ERROR = 'socket/AUTH_ERROR';
export var DISCONNECTED = 'socket/DISCONNECTED';
export var DEAUTHENTICATE = 'socket/DEAUTHENTICATE';
export var SUBSCRIBE_REQUEST = 'socket/SUBSCRIBE_REQUEST';
export var SUBSCRIBE_SUCCESS = 'socket/SUBSCRIBE_SUCCESS';
export var SUBSCRIBE_ERROR = 'socket/SUBSCRIBE_ERROR';
export var UNSUBSCRIBE = 'socket/UNSUBSCRIBE';
export var EMIT = 'socket/EMIT';
