// @redux-devtools/app/lib/cjs/containers/monitors/InspectorWrapper/ChartTab.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _reactRedux = require("react-redux");

var _styledComponents = require("styled-components");

var _d3StateVisualizer = require("d3-state-visualizer");

var _ChartMonitorWrapper = require("../ChartMonitorWrapper");

var _actions = require("../../../actions");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var style = {
  width: '100%',
  height: '100%'
};

var ChartTab = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(ChartTab, _Component);

  var _super = _createSuper(ChartTab);

  function ChartTab() {
    var _this;

    (0, _classCallCheck2.default)(this, ChartTab);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "getRef", function (node) {
      _this.node = node;
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onClickText", function (data) {
      var inspectedStatePath = [];
      (0, _ChartMonitorWrapper.getPath)(data, inspectedStatePath);

      _this.props.updateMonitorState({
        inspectedStatePath: inspectedStatePath,
        subTabName: data.children ? 'Chart' : 'Tree'
      });
    });
    return _this;
  }

  (0, _createClass2.default)(ChartTab, [{
    key: "shouldComponentUpdate",
    value: // eslint-disable-next-line @typescript-eslint/ban-types
    function shouldComponentUpdate() {
      return false;
    }
  }, {
    key: "componentDidMount",
    value: function componentDidMount() {
      this.createChart(this.props);
    }
  }, {
    key: "UNSAFE_componentWillReceiveProps",
    value: function UNSAFE_componentWillReceiveProps(nextProps) {
      if (this.props.theme.scheme !== nextProps.theme.scheme || nextProps.theme.light !== this.props.theme.light) {
        this.node.innerHTML = '';
        this.createChart(nextProps);
      } else if (nextProps.data !== this.props.data) {
        // eslint-disable-next-line @typescript-eslint/ban-types
        this.renderChart(nextProps.data);
      }
    }
  }, {
    key: "createChart",
    value: function createChart(props) {
      this.renderChart = (0, _d3StateVisualizer.tree)(this.node, this.getChartTheme(props.theme)); // eslint-disable-next-line @typescript-eslint/ban-types

      this.renderChart(props.data);
    }
  }, {
    key: "getChartTheme",
    value: function getChartTheme(theme) {
      return {
        heightBetweenNodesCoeff: 1,
        widthBetweenNodesCoeff: 1.3,
        tooltipOptions: {
          style: {
            color: theme.base06,
            'background-color': theme.base01,
            opacity: '0.9',
            'border-radius': '5px',
            padding: '5px'
          },
          offset: {
            left: 30,
            top: 10
          },
          indentationSize: 2
        },
        style: {
          width: '100%',
          height: '100%',
          node: {
            colors: {
              default: theme.base0B,
              collapsed: theme.base0B,
              parent: theme.base0E
            },
            radius: 7
          },
          text: {
            colors: {
              default: theme.base0D,
              hover: theme.base06
            }
          }
        },
        onClickText: this.onClickText
      };
    }
  }, {
    key: "render",
    value: function render() {
      return /*#__PURE__*/_react.default.createElement("div", {
        style: style,
        ref: this.getRef
      });
    }
  }]);
  return ChartTab;
}(_react.Component);

var actionCreators = {
  updateMonitorState: _actions.updateMonitorState
};
var ConnectedChartTab = (0, _reactRedux.connect)(null, actionCreators)(ChartTab);

var _default = (0, _styledComponents.withTheme)(ConnectedChartTab);

exports.default = _default;
