// @redux-devtools/ui/lib/cjs/Editor/Editor.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.WithTabs = exports.Default = void 0;

var _react = _interopRequireDefault(require("react"));

var _ = _interopRequireDefault(require("./"));

var _WithTabs = _interopRequireDefault(require("./WithTabs"));

var value = "\nvar themes = [];\n\nfunction getThemes() {\n  return themes;\n}\n";
var _default = {
  title: 'Editor',
  component: _.default
};
exports.default = _default;

var Template = function Template(args) {
  return /*#__PURE__*/_react.default.createElement(_.default, args);
};

var Default = Template.bind({});
exports.Default = Default;
Default.args = {
  value: value,
  lineNumbers: true,
  lineWrapping: false,
  foldGutter: true,
  readOnly: false,
  autofocus: true
};
Default.argTypes = {
  autofocus: {
    control: {
      disable: true
    }
  },
  mode: {
    control: {
      disable: true
    }
  },
  theme: {
    control: {
      disable: true
    }
  },
  onChange: {
    control: {
      disable: true
    }
  }
};

var WithTabsTemplate = function WithTabsTemplate(args) {
  return /*#__PURE__*/_react.default.createElement(_WithTabs.default, args);
};

var WithTabs = WithTabsTemplate.bind({});
exports.WithTabs = WithTabs;
WithTabs.args = {
  lineNumbers: true,
  position: 'left'
};
WithTabs.argTypes = {
  value: {
    control: {
      disable: true
    }
  },
  mode: {
    control: {
      disable: true
    }
  },
  lineWrapping: {
    control: {
      disable: true
    }
  },
  readOnly: {
    control: {
      disable: true
    }
  },
  theme: {
    control: {
      disable: true
    }
  },
  foldGutter: {
    control: {
      disable: true
    }
  },
  autofocus: {
    control: {
      disable: true
    }
  },
  onChange: {
    control: {
      disable: true
    }
  }
};
