// @redux-devtools/ui/lib/cjs/SegmentedControl/SegmentedControl.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _ = _interopRequireDefault(require("./"));

var _excluded = ["values"];

var _templateObject;

var Container = _styledComponents.default.div(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2.default)(["\n  display: flex;\n  height: 100%;\n  width: 100%;\n  justify-content: center;\n  align-items: center;\n"])));

var _default = {
  title: 'SegmentedControl',
  component: _.default
}; // eslint-disable-next-line react/prop-types

exports.default = _default;

var Template = function Template(_ref) {
  var values = _ref.values,
      args = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  return /*#__PURE__*/_react.default.createElement(Container, null, /*#__PURE__*/_react.default.createElement(_.default, (0, _extends2.default)({
    values: ['Button1', 'Button2', 'Button3']
  }, args)));
};

var Default = Template.bind({});
exports.Default = Default;
Default.args = {
  selected: 'Button1',
  disabled: false
};
Default.argTypes = {
  values: {
    control: {
      disable: true
    }
  },
  onClick: {
    control: {
      disable: true
    }
  },
  theme: {
    control: {
      disable: true
    }
  }
};
