// @redux-devtools/app/lib/cjs/components/Settings/Connection.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Connection = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _reactRedux = require("react-redux");

var _ui = require("@redux-devtools/ui");

var _actions = require("../../actions");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var defaultSchema = {
  type: 'object',
  required: [],
  properties: {
    type: {
      title: 'Connection settings (for getting reports and remote debugging)',
      type: 'string',
      enum: ['disabled', 'custom'],
      enumNames: ['no remote connection', 'use local (custom) server']
    },
    hostname: {
      type: 'string'
    },
    port: {
      type: 'number'
    },
    secure: {
      type: 'boolean'
    }
  }
};
var uiSchema = {
  type: {
    'ui:widget': 'radio'
  }
};

var Connection = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Connection, _Component);

  var _super = _createSuper(Connection);

  function Connection() {
    var _this;

    (0, _classCallCheck2.default)(this, Connection);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "setFormData", function (type, changed) {
      var schema;

      if (type !== 'custom') {
        schema = {
          type: 'object',
          properties: {
            type: defaultSchema.properties.type
          }
        };
      } else {
        schema = defaultSchema;
      }

      return {
        formData: _objectSpread({
          type: type
        }, _this.props.options),
        type: type,
        schema: schema,
        changed: changed
      };
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "state", _this.setFormData(_this.props.type));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleSave", function (data) {
      _this.props.saveSettings(data.formData);

      _this.setState({
        changed: false
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleChange", function (data) {
      var formData = data.formData;
      var type = formData.type;

      if (type !== _this.state.type) {
        _this.setState(_this.setFormData(type, true));
      } else if (!_this.state.changed) {
        _this.setState({
          changed: true,
          formData: formData
        });
      }
    });
    return _this;
  }

  (0, _createClass2.default)(Connection, [{
    key: "shouldComponentUpdate",
    value: function shouldComponentUpdate(nextProps, nextState) {
      return this.state !== nextState;
    }
  }, {
    key: "UNSAFE_componentWillReceiveProps",
    value: function UNSAFE_componentWillReceiveProps(nextProps) {
      if (this.props.options !== nextProps.options) {
        this.setState({
          formData: _objectSpread(_objectSpread({}, nextProps.options), {}, {
            type: nextProps.type
          })
        });
      }
    }
  }, {
    key: "render",
    value: function render() {
      var type = this.state.type || 'disabled';
      var changed = this.state.changed;
      var disabled = type === 'disabled';
      return /*#__PURE__*/_react.default.createElement(_ui.Container, null, /*#__PURE__*/_react.default.createElement(_ui.Form, {
        primaryButton: changed,
        noSubmit: disabled && !changed,
        submitText: disabled ? 'Disconnect' : 'Connect',
        formData: this.state.formData,
        schema: this.state.schema,
        uiSchema: uiSchema,
        onChange: this.handleChange,
        onSubmit: this.handleSave
      }));
    }
  }]);
  return Connection;
}(_react.Component);

exports.Connection = Connection;

var mapStateToProps = function mapStateToProps(state) {
  return state.connection;
};

var actionCreators = {
  saveSettings: _actions.saveSocketSettings
};

var _default = (0, _reactRedux.connect)(mapStateToProps, actionCreators)(Connection);

exports.default = _default;
