// @redux-devtools/ui/lib/esm/Select/Select.stories.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";
var _excluded = ["value"];

var _templateObject;

import React from 'react';
import styled from 'styled-components';
import Select from './';
import { options } from './options';
var Container = styled.div(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  display: flex;\n  height: 100%;\n  width: 100%;\n  justify-content: center;\n  align-items: center;\n\n  > div {\n    width: 90%;\n  }\n"])));
export default {
  title: 'Select',
  component: Select
};

// eslint-disable-next-line react/prop-types
var Template = function Template(_ref) {
  var value = _ref.value,
      args = _objectWithoutProperties(_ref, _excluded);

  return /*#__PURE__*/React.createElement(Container, null, /*#__PURE__*/React.createElement(Select, _extends({
    options: options,
    value: options.filter(function (option) {
      return option.value === value;
    })
  }, args)));
};

export var Default = Template.bind({});
Default.args = {
  value: 'one',
  maxMenuHeight: 300,
  isClearable: false,
  isDisabled: false,
  isLoading: false,
  isMulti: false,
  isSearchable: true,
  menuPlacement: 'bottom'
};
Default.argTypes = {
  onChange: {
    action: 'selected'
  }
};
