// @redux-devtools/app/lib/esm/utils/getMonitor.js
import React from 'react';
import { LogMonitor } from '@redux-devtools/log-monitor';
import ChartMonitorWrapper from '../containers/monitors/ChartMonitorWrapper';
import InspectorWrapper from '../containers/monitors/InspectorWrapper';
import { RtkQueryMonitor } from '@redux-devtools/rtk-query-monitor';
export var monitors = [{
  value: 'InspectorMonitor',
  name: 'Inspector'
}, {
  value: 'LogMonitor',
  name: 'Log monitor'
}, {
  value: 'ChartMonitor',
  name: 'Chart'
}, {
  value: 'RtkQueryMonitor',
  name: 'RTK Query'
}];
export default function getMonitor(_ref) {
  var monitor = _ref.monitor;

  switch (monitor) {
    case 'LogMonitor':
      return /*#__PURE__*/React.createElement(LogMonitor, {
        preserveScrollTop: false,
        hideMainButtons: true,
        markStateDiff: true
      });

    case 'ChartMonitor':
      return /*#__PURE__*/React.createElement(ChartMonitorWrapper, null);

    case 'RtkQueryMonitor':
      return /*#__PURE__*/React.createElement(RtkQueryMonitor, null);

    default:
      return /*#__PURE__*/React.createElement(InspectorWrapper, null);
  }
}
