// @redux-devtools/inspector-monitor/lib/cjs/tabs/ActionTab.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactJsonTree = require("react-json-tree");

var _getItemString2 = _interopRequireDefault(require("./getItemString"));

var _getJsonTreeTheme = _interopRequireDefault(require("./getJsonTreeTheme"));

var ActionTab = function ActionTab(_ref) {
  var action = _ref.action,
      styling = _ref.styling,
      base16Theme = _ref.base16Theme,
      invertTheme = _ref.invertTheme,
      labelRenderer = _ref.labelRenderer,
      dataTypeKey = _ref.dataTypeKey,
      isWideLayout = _ref.isWideLayout;
  return /*#__PURE__*/_react.default.createElement(_reactJsonTree.JSONTree, {
    labelRenderer: labelRenderer,
    theme: (0, _getJsonTreeTheme.default)(base16Theme),
    data: action,
    getItemString: function getItemString(type, data) {
      return (0, _getItemString2.default)(styling, type, data, dataTypeKey, isWideLayout);
    },
    invertTheme: invertTheme,
    hideRoot: true
  });
};

ActionTab.propTypes = {
  action: _propTypes.default.any.isRequired,
  styling: _propTypes.default.func.isRequired,
  base16Theme: _propTypes.default.any.isRequired,
  invertTheme: _propTypes.default.bool.isRequired,
  labelRenderer: _propTypes.default.func.isRequired,
  dataTypeKey: _propTypes.default.string,
  isWideLayout: _propTypes.default.bool.isRequired
};
var _default = ActionTab;
exports.default = _default;
