// @redux-devtools/inspector-monitor/lib/cjs/tabs/DiffTab.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _JSONDiff = _interopRequireDefault(require("./JSONDiff"));

var DiffTab = function DiffTab(_ref) {
  var delta = _ref.delta,
      styling = _ref.styling,
      base16Theme = _ref.base16Theme,
      invertTheme = _ref.invertTheme,
      labelRenderer = _ref.labelRenderer,
      isWideLayout = _ref.isWideLayout,
      dataTypeKey = _ref.dataTypeKey;
  return /*#__PURE__*/_react.default.createElement(_JSONDiff.default, {
    delta: delta,
    styling: styling,
    base16Theme: base16Theme,
    invertTheme: invertTheme,
    labelRenderer: labelRenderer,
    isWideLayout: isWideLayout,
    dataTypeKey: dataTypeKey
  });
};

DiffTab.propTypes = {
  delta: _propTypes.default.any,
  styling: _propTypes.default.func.isRequired,
  base16Theme: _propTypes.default.any.isRequired,
  invertTheme: _propTypes.default.bool.isRequired,
  labelRenderer: _propTypes.default.func.isRequired,
  isWideLayout: _propTypes.default.bool.isRequired,
  dataTypeKey: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.symbol])
};
var _default = DiffTab;
exports.default = _default;
