// @redux-devtools/chart-monitor/lib/cjs/Chart.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _d3StateVisualizer = require("d3-state-visualizer");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var wrapperStyle = {
  width: '100%',
  height: '100%'
};

var Chart = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Chart, _Component);

  var _super = _createSuper(Chart);

  function Chart() {
    var _this;

    (0, _classCallCheck2.default)(this, Chart);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "divRef", /*#__PURE__*/(0, _react.createRef)());
    return _this;
  }

  (0, _createClass2.default)(Chart, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var _this$props = this.props,
          select = _this$props.select,
          state = _this$props.state,
          defaultIsVisible = _this$props.defaultIsVisible;
      this.renderChart = (0, _d3StateVisualizer.tree)(this.divRef.current, this.props);

      if (defaultIsVisible) {
        // eslint-disable-next-line @typescript-eslint/ban-types
        this.renderChart(select(state));
      }
    }
  }, {
    key: "UNSAFE_componentWillReceiveProps",
    value: function UNSAFE_componentWillReceiveProps(nextProps) {
      var state = nextProps.state,
          select = nextProps.select,
          monitorState = nextProps.monitorState;

      if (monitorState.isVisible !== false) {
        // eslint-disable-next-line @typescript-eslint/ban-types
        this.renderChart(select(state));
      }
    }
  }, {
    key: "render",
    value: function render() {
      return /*#__PURE__*/_react.default.createElement("div", {
        style: wrapperStyle,
        ref: this.divRef
      });
    }
  }]);
  return Chart;
}(_react.Component);

(0, _defineProperty2.default)(Chart, "propTypes", {
  state: _propTypes.default.object,
  rootKeyName: _propTypes.default.string,
  pushMethod: _propTypes.default.oneOf(['push', 'unshift']),
  tree: _propTypes.default.shape({
    name: _propTypes.default.string,
    children: _propTypes.default.array
  }),
  id: _propTypes.default.string,
  style: _propTypes.default.shape({
    node: _propTypes.default.shape({
      colors: _propTypes.default.shape({
        default: _propTypes.default.string,
        parent: _propTypes.default.string,
        collapsed: _propTypes.default.string
      }),
      radius: _propTypes.default.number
    }),
    text: _propTypes.default.shape({
      colors: _propTypes.default.shape({
        default: _propTypes.default.string,
        hover: _propTypes.default.string
      })
    }),
    link: _propTypes.default.object
  }),
  size: _propTypes.default.number,
  aspectRatio: _propTypes.default.number,
  margin: _propTypes.default.shape({
    top: _propTypes.default.number,
    right: _propTypes.default.number,
    bottom: _propTypes.default.number,
    left: _propTypes.default.number
  }),
  isSorted: _propTypes.default.bool,
  heightBetweenNodesCoeff: _propTypes.default.number,
  widthBetweenNodesCoeff: _propTypes.default.number,
  transitionDuration: _propTypes.default.number,
  onClickText: _propTypes.default.func,
  tooltipOptions: _propTypes.default.shape({
    disabled: _propTypes.default.bool,
    left: _propTypes.default.number,
    top: _propTypes.default.number,
    offset: _propTypes.default.shape({
      left: _propTypes.default.number,
      top: _propTypes.default.number
    }),
    indentationSize: _propTypes.default.number,
    style: _propTypes.default.object
  })
});
var _default = Chart;
exports.default = _default;
