// @redux-devtools/app/lib/cjs/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  Root: true,
  DispatcherButton: true,
  ExportButton: true,
  ImportButton: true,
  PrintButton: true,
  SliderButton: true,
  Header: true,
  MonitorSelector: true,
  Settings: true,
  TopButtons: true,
  DevTools: true,
  Dispatcher: true,
  SliderMonitor: true
};
Object.defineProperty(exports, "DevTools", {
  enumerable: true,
  get: function get() {
    return _DevTools.default;
  }
});
Object.defineProperty(exports, "Dispatcher", {
  enumerable: true,
  get: function get() {
    return _Dispatcher.default;
  }
});
Object.defineProperty(exports, "DispatcherButton", {
  enumerable: true,
  get: function get() {
    return _DispatcherButton.default;
  }
});
Object.defineProperty(exports, "ExportButton", {
  enumerable: true,
  get: function get() {
    return _ExportButton.default;
  }
});
Object.defineProperty(exports, "Header", {
  enumerable: true,
  get: function get() {
    return _Header.default;
  }
});
Object.defineProperty(exports, "ImportButton", {
  enumerable: true,
  get: function get() {
    return _ImportButton.default;
  }
});
Object.defineProperty(exports, "MonitorSelector", {
  enumerable: true,
  get: function get() {
    return _MonitorSelector.default;
  }
});
Object.defineProperty(exports, "PrintButton", {
  enumerable: true,
  get: function get() {
    return _PrintButton.default;
  }
});
exports.Root = void 0;
Object.defineProperty(exports, "Settings", {
  enumerable: true,
  get: function get() {
    return _Settings.default;
  }
});
Object.defineProperty(exports, "SliderButton", {
  enumerable: true,
  get: function get() {
    return _SliderButton.default;
  }
});
Object.defineProperty(exports, "SliderMonitor", {
  enumerable: true,
  get: function get() {
    return _Slider.default;
  }
});
Object.defineProperty(exports, "TopButtons", {
  enumerable: true,
  get: function get() {
    return _TopButtons.default;
  }
});

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _react = _interopRequireWildcard(require("react"));

var _reactRedux = require("react-redux");

var _react2 = require("redux-persist/integration/react");

var _configureStore2 = _interopRequireDefault(require("./store/configureStore"));

var _socketActionTypes = require("./constants/socketActionTypes");

Object.keys(_socketActionTypes).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _socketActionTypes[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _socketActionTypes[key];
    }
  });
});

var _App = _interopRequireDefault(require("./containers/App"));

var _actions = require("./actions");

Object.keys(_actions).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _actions[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _actions[key];
    }
  });
});

var _DispatcherButton = _interopRequireDefault(require("./components/buttons/DispatcherButton"));

var _ExportButton = _interopRequireDefault(require("./components/buttons/ExportButton"));

var _ImportButton = _interopRequireDefault(require("./components/buttons/ImportButton"));

var _PrintButton = _interopRequireDefault(require("./components/buttons/PrintButton"));

var _SliderButton = _interopRequireDefault(require("./components/buttons/SliderButton"));

var _Header = _interopRequireDefault(require("./components/Header"));

var _MonitorSelector = _interopRequireDefault(require("./components/MonitorSelector"));

var _Settings = _interopRequireDefault(require("./components/Settings"));

var _TopButtons = _interopRequireDefault(require("./components/TopButtons"));

var _DevTools = _interopRequireDefault(require("./containers/DevTools"));

var _Dispatcher = _interopRequireDefault(require("./containers/monitors/Dispatcher"));

var _Slider = _interopRequireDefault(require("./containers/monitors/Slider"));

var _actionTypes = require("./constants/actionTypes");

Object.keys(_actionTypes).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _actionTypes[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _actionTypes[key];
    }
  });
});

var _api = require("./middlewares/api");

Object.keys(_api).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _api[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _api[key];
    }
  });
});

var _exportState = require("./middlewares/exportState");

Object.keys(_exportState).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _exportState[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _exportState[key];
    }
  });
});

var _reducers = require("./reducers");

Object.keys(_reducers).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _reducers[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _reducers[key];
    }
  });
});

var _connection = require("./reducers/connection");

Object.keys(_connection).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _connection[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _connection[key];
    }
  });
});

var _instances = require("./reducers/instances");

Object.keys(_instances).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _instances[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _instances[key];
    }
  });
});

var _monitor = require("./reducers/monitor");

Object.keys(_monitor).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _monitor[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _monitor[key];
    }
  });
});

var _notification = require("./reducers/notification");

Object.keys(_notification).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _notification[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _notification[key];
    }
  });
});

var _reports = require("./reducers/reports");

Object.keys(_reports).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _reports[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _reports[key];
    }
  });
});

var _section = require("./reducers/section");

Object.keys(_section).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _section[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _section[key];
    }
  });
});

var _socket = require("./reducers/socket");

Object.keys(_socket).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _socket[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _socket[key];
    }
  });
});

var _theme = require("./reducers/theme");

Object.keys(_theme).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _theme[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _theme[key];
    }
  });
});

var _monitorActions = require("./utils/monitorActions");

Object.keys(_monitorActions).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _monitorActions[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _monitorActions[key];
    }
  });
});

var _stringifyJSON = require("./utils/stringifyJSON");

Object.keys(_stringifyJSON).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _stringifyJSON[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _stringifyJSON[key];
    }
  });
});

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var Root = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Root, _Component);

  var _super = _createSuper(Root);

  function Root() {
    (0, _classCallCheck2.default)(this, Root);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(Root, [{
    key: "UNSAFE_componentWillMount",
    value: function UNSAFE_componentWillMount() {
      var _configureStore = (0, _configureStore2.default)(function (store) {
        if (store.getState().connection.type !== 'disabled') {
          store.dispatch({
            type: _socketActionTypes.CONNECT_REQUEST
          });
        }
      }),
          store = _configureStore.store,
          persistor = _configureStore.persistor;

      this.store = store;
      this.persistor = persistor;
    }
  }, {
    key: "render",
    value: function render() {
      if (!this.store) return null;
      return /*#__PURE__*/_react.default.createElement(_reactRedux.Provider, {
        store: this.store
      }, /*#__PURE__*/_react.default.createElement(_react2.PersistGate, {
        loading: null,
        persistor: this.persistor
      }, /*#__PURE__*/_react.default.createElement(_App.default, null)));
    }
  }]);
  return Root;
}(_react.Component);

exports.Root = Root;
