// @redux-devtools/app/lib/cjs/utils/getMonitor.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = getMonitor;
exports.monitors = void 0;

var _react = _interopRequireDefault(require("react"));

var _logMonitor = require("@redux-devtools/log-monitor");

var _ChartMonitorWrapper = _interopRequireDefault(require("../containers/monitors/ChartMonitorWrapper"));

var _InspectorWrapper = _interopRequireDefault(require("../containers/monitors/InspectorWrapper"));

var _rtkQueryMonitor = require("@redux-devtools/rtk-query-monitor");

var monitors = [{
  value: 'InspectorMonitor',
  name: 'Inspector'
}, {
  value: 'LogMonitor',
  name: 'Log monitor'
}, {
  value: 'ChartMonitor',
  name: 'Chart'
}, {
  value: 'RtkQueryMonitor',
  name: 'RTK Query'
}];
exports.monitors = monitors;

function getMonitor(_ref) {
  var monitor = _ref.monitor;

  switch (monitor) {
    case 'LogMonitor':
      return /*#__PURE__*/_react.default.createElement(_logMonitor.LogMonitor, {
        preserveScrollTop: false,
        hideMainButtons: true,
        markStateDiff: true
      });

    case 'ChartMonitor':
      return /*#__PURE__*/_react.default.createElement(_ChartMonitorWrapper.default, null);

    case 'RtkQueryMonitor':
      return /*#__PURE__*/_react.default.createElement(_rtkQueryMonitor.RtkQueryMonitor, null);

    default:
      return /*#__PURE__*/_react.default.createElement(_InspectorWrapper.default, null);
  }
}
