// @redux-devtools/slider-monitor/lib/cjs/SliderButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _ui = require("@redux-devtools/ui");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var SliderButton = /*#__PURE__*/function (_ref) {
  (0, _inherits2.default)(SliderButton, _ref);

  var _super = _createSuper(SliderButton);

  function SliderButton() {
    var _this;

    (0, _classCallCheck2.default)(this, SliderButton);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "renderPauseButton", function () {
      return /*#__PURE__*/_react.default.createElement(_ui.Button, {
        onClick: _this.props.onClick,
        title: "Pause",
        size: "small",
        disabled: _this.props.disabled,
        theme: _this.props.theme
      }, /*#__PURE__*/_react.default.createElement("svg", {
        viewBox: "0 0 24 24",
        preserveAspectRatio: "xMidYMid meet",
        style: _this.iconStyle()
      }, /*#__PURE__*/_react.default.createElement("g", null, /*#__PURE__*/_react.default.createElement("path", {
        d: "M6 19h4V5H6v14zm8-14v14h4V5h-4z"
      }))));
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "renderStepButton", function (direction) {
      var isLeft = direction === 'left';
      var d = isLeft ? 'M15.41 16.09l-4.58-4.59 4.58-4.59-1.41-1.41-6 6 6 6z' : 'M8.59 16.34l4.58-4.59-4.58-4.59 1.41-1.41 6 6-6 6z';
      return /*#__PURE__*/_react.default.createElement(_ui.Button, {
        size: "small",
        title: isLeft ? 'Go back' : 'Go forward',
        onClick: _this.props.onClick,
        disabled: _this.props.disabled,
        theme: _this.props.theme
      }, /*#__PURE__*/_react.default.createElement("svg", {
        viewBox: "0 0 24 24",
        preserveAspectRatio: "xMidYMid meet",
        style: _this.iconStyle()
      }, /*#__PURE__*/_react.default.createElement("g", null, /*#__PURE__*/_react.default.createElement("path", {
        d: d
      }))));
    });
    return _this;
  }

  (0, _createClass2.default)(SliderButton, [{
    key: "iconStyle",
    value: function iconStyle() {
      return {
        cursor: 'hand',
        fill: this.props.theme.base06,
        width: '1.8rem',
        height: '1.8rem'
      };
    }
  }, {
    key: "renderPlayButton",
    value: function renderPlayButton() {
      return /*#__PURE__*/_react.default.createElement(_ui.Button, {
        onClick: this.props.onClick,
        title: "Play",
        size: "small",
        disabled: this.props.disabled,
        theme: this.props.theme
      }, /*#__PURE__*/_react.default.createElement("svg", {
        viewBox: "0 0 24 24",
        preserveAspectRatio: "xMidYMid meet",
        style: this.iconStyle()
      }, /*#__PURE__*/_react.default.createElement("g", null, /*#__PURE__*/_react.default.createElement("path", {
        d: "M8 5v14l11-7z"
      }))));
    }
  }, {
    key: "render",
    value: function render() {
      switch (this.props.type) {
        case 'play':
          return this.renderPlayButton();

        case 'pause':
          return this.renderPauseButton();

        case 'stepLeft':
          return this.renderStepButton('left');

        case 'stepRight':
          return this.renderStepButton('right');

        default:
          return null;
      }
    }
  }]);
  return SliderButton;
}(_react.PureComponent || _react.Component);

exports.default = SliderButton;
(0, _defineProperty2.default)(SliderButton, "propTypes", {
  theme: _propTypes.default.object,
  type: _propTypes.default.string,
  disabled: _propTypes.default.bool,
  onClick: _propTypes.default.func
});
