// @redux-devtools/app/lib/esm/middlewares/exportState.js
import { stringifyJSON } from '../utils/stringifyJSON';
import { UPDATE_STATE, LIFTED_ACTION, EXPORT } from '../constants/actionTypes';
import { getActiveInstance } from '../reducers/instances';
var toExport;

function download(state) {
  var blob = new Blob([state], {
    type: 'octet/stream'
  });
  var href = window.URL.createObjectURL(blob);
  var a = document.createElement('a');
  a.style.display = 'none';
  a.download = 'state.json';
  a.href = href;
  document.body.appendChild(a);
  a.click();
  setTimeout(function () {
    document.body.removeChild(a);
    window.URL.revokeObjectURL(href);
  }, 0);
}

export var exportStateMiddleware = function exportStateMiddleware(store) {
  return function (next) {
    return function (action) {
      var result = next(action);

      if (toExport && action.type === UPDATE_STATE && action.request.type === 'EXPORT') {
        var request = action.request;
        var id = request.instanceId || request.id;

        if (id === toExport) {
          toExport = undefined;
          download(JSON.stringify({
            payload: request.payload,
            preloadedState: request.committedState
          }, null, '\t'));
        }
      } else if (action.type === EXPORT) {
        var instances = store.getState().instances;
        var instanceId = getActiveInstance(instances);
        var options = instances.options[instanceId];

        if (options.features.export === true) {
          download(stringifyJSON(instances.states[instanceId], options.serialize));
        } else {
          toExport = instanceId;
          next({
            type: LIFTED_ACTION,
            message: 'EXPORT',
            toExport: true
          });
        }
      }

      return result;
    };
  };
};
