// @redux-devtools/rtk-query-monitor/lib/cjs/utils/comparators.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.QueryComparators = void 0;
exports.compareJSONPrimitive = compareJSONPrimitive;
exports.sortQueryOptions = exports.queryComparators = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _query = require("@reduxjs/toolkit/query");

var _mapStatusToFactor, _queryComparators;

var QueryComparators;
exports.QueryComparators = QueryComparators;

(function (QueryComparators) {
  QueryComparators["fulfilledTimeStamp"] = "timestamp";
  QueryComparators["queryKey"] = "key";
  QueryComparators["status"] = "status";
  QueryComparators["endpointName"] = "endpointName";
  QueryComparators["apiReducerPath"] = "apiReducerPath";
})(QueryComparators || (exports.QueryComparators = QueryComparators = {}));

var sortQueryOptions = [{
  label: 'last updated',
  value: QueryComparators.fulfilledTimeStamp
}, {
  label: 'query key',
  value: QueryComparators.queryKey
}, {
  label: 'status',
  value: QueryComparators.status
}, {
  label: 'endpoint',
  value: QueryComparators.endpointName
}, {
  label: 'reducerPath',
  value: QueryComparators.apiReducerPath
}];
exports.sortQueryOptions = sortQueryOptions;

function sortQueryByFulfilled(thisQueryInfo, thatQueryInfo) {
  var _thisQueryInfo$state$, _thatQueryInfo$state$;

  var thisFulfilled = (_thisQueryInfo$state$ = thisQueryInfo.state.fulfilledTimeStamp) !== null && _thisQueryInfo$state$ !== void 0 ? _thisQueryInfo$state$ : -1;
  var thatFulfilled = (_thatQueryInfo$state$ = thatQueryInfo.state.fulfilledTimeStamp) !== null && _thatQueryInfo$state$ !== void 0 ? _thatQueryInfo$state$ : -1;
  return thisFulfilled - thatFulfilled;
}

var mapStatusToFactor = (_mapStatusToFactor = {}, (0, _defineProperty2.default)(_mapStatusToFactor, _query.QueryStatus.uninitialized, 1), (0, _defineProperty2.default)(_mapStatusToFactor, _query.QueryStatus.pending, 2), (0, _defineProperty2.default)(_mapStatusToFactor, _query.QueryStatus.rejected, 3), (0, _defineProperty2.default)(_mapStatusToFactor, _query.QueryStatus.fulfilled, 4), _mapStatusToFactor);

function sortQueryByStatus(thisQueryInfo, thatQueryInfo) {
  var thisTerm = mapStatusToFactor[thisQueryInfo.state.status] || -1;
  var thatTerm = mapStatusToFactor[thatQueryInfo.state.status] || -1;
  return thisTerm - thatTerm;
}

function compareJSONPrimitive(a, b) {
  if (a === b) {
    return 0;
  }

  return a > b ? 1 : -1;
}

function sortByQueryKey(thisQueryInfo, thatQueryInfo) {
  return compareJSONPrimitive(thisQueryInfo.queryKey, thatQueryInfo.queryKey);
}

function sortQueryByEndpointName(thisQueryInfo, thatQueryInfo) {
  var _thisQueryInfo$state$2, _thatQueryInfo$state$2;

  var thisEndpointName = (_thisQueryInfo$state$2 = thisQueryInfo.state.endpointName) !== null && _thisQueryInfo$state$2 !== void 0 ? _thisQueryInfo$state$2 : '';
  var thatEndpointName = (_thatQueryInfo$state$2 = thatQueryInfo.state.endpointName) !== null && _thatQueryInfo$state$2 !== void 0 ? _thatQueryInfo$state$2 : '';
  return compareJSONPrimitive(thisEndpointName, thatEndpointName);
}

function sortByApiReducerPath(thisQueryInfo, thatQueryInfo) {
  return compareJSONPrimitive(thisQueryInfo.reducerPath, thatQueryInfo.reducerPath);
}

var queryComparators = (_queryComparators = {}, (0, _defineProperty2.default)(_queryComparators, QueryComparators.fulfilledTimeStamp, sortQueryByFulfilled), (0, _defineProperty2.default)(_queryComparators, QueryComparators.status, sortQueryByStatus), (0, _defineProperty2.default)(_queryComparators, QueryComparators.endpointName, sortQueryByEndpointName), (0, _defineProperty2.default)(_queryComparators, QueryComparators.queryKey, sortByQueryKey), (0, _defineProperty2.default)(_queryComparators, QueryComparators.apiReducerPath, sortByApiReducerPath), _queryComparators);
exports.queryComparators = queryComparators;
