// @redux-devtools/rtk-query-monitor/lib/cjs/utils/filters.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.queryListFilters = exports.filterQueryOptions = exports.QueryFilters = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _queryListFilters;

var QueryFilters;
exports.QueryFilters = QueryFilters;

(function (QueryFilters) {
  QueryFilters["queryKey"] = "query key";
  QueryFilters["reducerPath"] = "reducerPath";
  QueryFilters["endpointName"] = "endpoint";
  QueryFilters["status"] = "status";
})(QueryFilters || (exports.QueryFilters = QueryFilters = {}));

function filterByQueryKey(regex, list) {
  if (!regex) {
    return list;
  }

  return list.filter(function (RtkResourceInfo) {
    return regex.test(RtkResourceInfo.queryKey);
  });
}

function filterByReducerPath(regex, list) {
  if (!regex) {
    return list;
  }

  return list.filter(function (RtkResourceInfo) {
    return regex.test(RtkResourceInfo.reducerPath);
  });
}

function filterByEndpointName(regex, list) {
  if (!regex) {
    return list;
  }

  return list.filter(function (RtkResourceInfo) {
    var _RtkResourceInfo$stat;

    return regex.test((_RtkResourceInfo$stat = RtkResourceInfo.state.endpointName) !== null && _RtkResourceInfo$stat !== void 0 ? _RtkResourceInfo$stat : 'undefined');
  });
}

function filterByStatus(regex, list) {
  if (!regex) {
    return list;
  }

  return list.filter(function (RtkResourceInfo) {
    return regex.test(RtkResourceInfo.state.status);
  });
}

var filterQueryOptions = [{
  label: 'query key',
  value: QueryFilters.queryKey
}, {
  label: 'reducerPath',
  value: QueryFilters.reducerPath
}, {
  label: 'status',
  value: QueryFilters.status
}, {
  label: 'endpoint',
  value: QueryFilters.endpointName
}];
exports.filterQueryOptions = filterQueryOptions;
var queryListFilters = (_queryListFilters = {}, (0, _defineProperty2.default)(_queryListFilters, QueryFilters.queryKey, filterByQueryKey), (0, _defineProperty2.default)(_queryListFilters, QueryFilters.endpointName, filterByEndpointName), (0, _defineProperty2.default)(_queryListFilters, QueryFilters.reducerPath, filterByReducerPath), (0, _defineProperty2.default)(_queryListFilters, QueryFilters.status, filterByStatus), _queryListFilters);
exports.queryListFilters = queryListFilters;
