// @redux-devtools/ui/lib/cjs/Dialog/Dialog.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _createStyledComponent = _interopRequireDefault(require("../utils/createStyledComponent"));

var styles = _interopRequireWildcard(require("./styles"));

var _Button = _interopRequireDefault(require("../Button"));

var _Form = _interopRequireDefault(require("../Form"));

var _excluded = ["modal", "open", "fullWidth", "title", "children", "actions", "noHeader", "noFooter", "submitText", "onDismiss"];

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var DialogWrapper = (0, _createStyledComponent.default)(styles);

function isForm(rest) {
  return rest.schema !== undefined;
}

var Dialog = /*#__PURE__*/function (_PureComponent) {
  (0, _inherits2.default)(Dialog, _PureComponent);

  var _super = _createSuper(Dialog);

  function Dialog() {
    var _this;

    (0, _classCallCheck2.default)(this, Dialog);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onSubmit", function () {
      if (_this.submitButton) _this.submitButton.click();else _this.props.onSubmit();
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "getFormButtonRef", function (node) {
      _this.submitButton = node;
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onKeyDown", function (e) {
      if (e.keyCode === 27
      /* esc */
      ) {
        e.preventDefault();

        _this.props.onDismiss(false);
      }
    });
    return _this;
  }

  (0, _createClass2.default)(Dialog, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          modal = _this$props.modal,
          open = _this$props.open,
          fullWidth = _this$props.fullWidth,
          title = _this$props.title,
          children = _this$props.children,
          actions = _this$props.actions,
          noHeader = _this$props.noHeader,
          noFooter = _this$props.noFooter,
          submitText = _this$props.submitText,
          onDismiss = _this$props.onDismiss,
          rest = (0, _objectWithoutProperties2.default)(_this$props, _excluded);
      var schema = rest.schema;
      return /*#__PURE__*/_react.default.createElement(DialogWrapper, {
        open: open,
        fullWidth: fullWidth,
        onKeyDown: this.onKeyDown,
        theme: rest.theme
      }, /*#__PURE__*/_react.default.createElement("div", {
        onClick: !modal ? onDismiss : undefined
      }), /*#__PURE__*/_react.default.createElement("div", null, !noHeader && /*#__PURE__*/_react.default.createElement("div", {
        className: "mc-dialog--header"
      }, /*#__PURE__*/_react.default.createElement("div", null, schema ? schema.title || title : title), !modal && /*#__PURE__*/_react.default.createElement("button", {
        onClick: onDismiss
      }, "\xD7")), /*#__PURE__*/_react.default.createElement("div", {
        className: "mc-dialog--body"
      }, children, isForm(rest) && /*#__PURE__*/_react.default.createElement(_Form.default, rest, !noFooter && /*#__PURE__*/_react.default.createElement("input", {
        type: "submit",
        ref: this.getFormButtonRef,
        className: "mc-dialog--hidden"
      }))), !noFooter && (actions ? /*#__PURE__*/_react.default.createElement("div", {
        className: "mc-dialog--footer"
      }, submitText ? [].concat((0, _toConsumableArray2.default)(actions), [/*#__PURE__*/_react.default.createElement(_Button.default, {
        key: "default-submit",
        primary: true,
        onClick: this.onSubmit
      }, submitText)]) : actions) : /*#__PURE__*/_react.default.createElement("div", {
        className: "mc-dialog--footer"
      }, /*#__PURE__*/_react.default.createElement(_Button.default, {
        onClick: onDismiss
      }, "Cancel"), /*#__PURE__*/_react.default.createElement(_Button.default, {
        primary: true,
        onClick: this.onSubmit
      }, submitText || 'Submit')))));
    }
  }]);
  return Dialog;
}(_react.PureComponent);

exports.default = Dialog;
(0, _defineProperty2.default)(Dialog, "propTypes", {
  open: _propTypes.default.bool,
  title: _propTypes.default.string,
  children: _propTypes.default.any,
  actions: _propTypes.default.node,
  submitText: _propTypes.default.string,
  fullWidth: _propTypes.default.bool,
  noHeader: _propTypes.default.bool,
  noFooter: _propTypes.default.bool,
  modal: _propTypes.default.bool,
  onDismiss: _propTypes.default.func,
  onSubmit: _propTypes.default.func,
  theme: _propTypes.default.object
});
