// @redux-devtools/ui/lib/cjs/Dialog/Dialog.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.WithForm = exports.Default = void 0;

var _react = _interopRequireDefault(require("react"));

var _ = _interopRequireDefault(require("./"));

var _schema = require("../Form/schema");

var _default = {
  title: 'Dialog',
  component: _.default
};
exports.default = _default;

var Template = function Template(args) {
  return /*#__PURE__*/_react.default.createElement(_.default, args);
};

var Default = Template.bind({});
exports.Default = Default;
Default.args = {
  title: 'Dialog Title',
  submitText: 'Submit!',
  open: true,
  noHeader: false,
  noFooter: false,
  modal: false,
  fullWidth: false,
  children: 'Hello Dialog!'
};
Default.argTypes = {
  actions: {
    control: {
      disable: true
    }
  },
  onDismiss: {
    control: {
      disable: true
    }
  },
  onSubmit: {
    control: {
      disable: true
    }
  },
  theme: {
    control: {
      disable: true
    }
  }
};
var WithForm = Template.bind({});
exports.WithForm = WithForm;
WithForm.args = {
  open: true,
  noHeader: false,
  noFooter: false,
  fullWidth: false,
  submitText: 'Submit!',
  formData: _schema.formData,
  schema: _schema.schema,
  uiSchema: _schema.uiSchema
};
WithForm.argTypes = {
  title: {
    control: {
      disable: true
    }
  },
  children: {
    control: {
      disable: true
    }
  },
  actions: {
    control: {
      disable: true
    }
  },
  modal: {
    control: {
      disable: true
    }
  },
  onDismiss: {
    control: {
      disable: true
    }
  },
  onSubmit: {
    control: {
      disable: true
    }
  },
  theme: {
    control: {
      disable: true
    }
  }
};
