// @redux-devtools/rtk-query-monitor/lib/esm/utils/filters.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";

var _queryListFilters;

export var QueryFilters;

(function (QueryFilters) {
  QueryFilters["queryKey"] = "query key";
  QueryFilters["reducerPath"] = "reducerPath";
  QueryFilters["endpointName"] = "endpoint";
  QueryFilters["status"] = "status";
})(QueryFilters || (QueryFilters = {}));

function filterByQueryKey(regex, list) {
  if (!regex) {
    return list;
  }

  return list.filter(function (RtkResourceInfo) {
    return regex.test(RtkResourceInfo.queryKey);
  });
}

function filterByReducerPath(regex, list) {
  if (!regex) {
    return list;
  }

  return list.filter(function (RtkResourceInfo) {
    return regex.test(RtkResourceInfo.reducerPath);
  });
}

function filterByEndpointName(regex, list) {
  if (!regex) {
    return list;
  }

  return list.filter(function (RtkResourceInfo) {
    var _RtkResourceInfo$stat;

    return regex.test((_RtkResourceInfo$stat = RtkResourceInfo.state.endpointName) !== null && _RtkResourceInfo$stat !== void 0 ? _RtkResourceInfo$stat : 'undefined');
  });
}

function filterByStatus(regex, list) {
  if (!regex) {
    return list;
  }

  return list.filter(function (RtkResourceInfo) {
    return regex.test(RtkResourceInfo.state.status);
  });
}

export var filterQueryOptions = [{
  label: 'query key',
  value: QueryFilters.queryKey
}, {
  label: 'reducerPath',
  value: QueryFilters.reducerPath
}, {
  label: 'status',
  value: QueryFilters.status
}, {
  label: 'endpoint',
  value: QueryFilters.endpointName
}];
export var queryListFilters = (_queryListFilters = {}, _defineProperty(_queryListFilters, QueryFilters.queryKey, filterByQueryKey), _defineProperty(_queryListFilters, QueryFilters.endpointName, filterByEndpointName), _defineProperty(_queryListFilters, QueryFilters.reducerPath, filterByReducerPath), _defineProperty(_queryListFilters, QueryFilters.status, filterByStatus), _queryListFilters);
