// @redux-devtools/app/lib/esm/components/Header.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _defineProperty from "@babel/runtime/helpers/defineProperty";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

import React, { Component } from 'react';
import { Tabs, Toolbar, Button, Divider } from '@redux-devtools/ui';
import { connect } from 'react-redux';
import { GoBook } from 'react-icons/go';
import { IoMdText } from 'react-icons/io';
import { TiSocialTwitter } from 'react-icons/ti';
import { TiHeartFullOutline } from 'react-icons/ti';
import { changeSection } from '../actions';
var tabs = [{
  name: 'Actions'
}, {
  name: 'Settings'
}];

var Header = /*#__PURE__*/function (_Component) {
  _inherits(Header, _Component);

  var _super = _createSuper(Header);

  function Header() {
    var _this;

    _classCallCheck(this, Header);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "openLink", function (url) {
      return function () {
        window.open(url);
      };
    });

    return _this;
  }

  _createClass(Header, [{
    key: "render",
    value: function render() {
      return /*#__PURE__*/React.createElement(Toolbar, {
        compact: true,
        noBorder: true,
        borderPosition: "bottom"
      }, /*#__PURE__*/React.createElement(Tabs, {
        main: true,
        collapsible: true,
        tabs: tabs,
        onClick: this.props.changeSection,
        selected: this.props.section || 'Actions'
      }), /*#__PURE__*/React.createElement(Divider, null), /*#__PURE__*/React.createElement(Button, {
        title: "Documentation",
        tooltipPosition: "bottom",
        onClick: this.openLink('https://github.com/reduxjs/redux-devtools/blob/main/README.md')
      }, /*#__PURE__*/React.createElement(GoBook, null)), /*#__PURE__*/React.createElement(Button, {
        title: "Feedback",
        tooltipPosition: "bottom",
        onClick: this.openLink('https://github.com/reduxjs/redux-devtools/discussions')
      }, /*#__PURE__*/React.createElement(IoMdText, null)), /*#__PURE__*/React.createElement(Button, {
        title: "Follow us",
        tooltipPosition: "bottom",
        onClick: this.openLink('https://twitter.com/NathanBierema')
      }, /*#__PURE__*/React.createElement(TiSocialTwitter, null)), /*#__PURE__*/React.createElement(Button, {
        title: "Support us",
        tooltipPosition: "bottom-left",
        onClick: this.openLink('https://github.com/sponsors/Methuselah96')
      }, /*#__PURE__*/React.createElement(TiHeartFullOutline, null)));
    }
  }]);

  return Header;
}(Component);

var actionCreators = {
  changeSection: changeSection
};
export default connect(null, actionCreators)(Header);
