// @redux-devtools/app/lib/cjs/components/TopButtons.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _core = require("@redux-devtools/core");

var _ui = require("@redux-devtools/ui");

var _RecordButton = _interopRequireDefault(require("./buttons/RecordButton"));

var _PersistButton = _interopRequireDefault(require("./buttons/PersistButton"));

var _LockButton = _interopRequireDefault(require("./buttons/LockButton"));

var _InstanceSelector = _interopRequireDefault(require("./InstanceSelector"));

var _SyncButton = _interopRequireDefault(require("./buttons/SyncButton"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

// eslint-disable-next-line @typescript-eslint/unbound-method
var reset = _core.ActionCreators.reset,
    rollback = _core.ActionCreators.rollback,
    commit = _core.ActionCreators.commit,
    sweep = _core.ActionCreators.sweep;

var TopButtons = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(TopButtons, _Component);

  var _super = _createSuper(TopButtons);

  function TopButtons() {
    var _this;

    (0, _classCallCheck2.default)(this, TopButtons);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleRollback", function () {
      _this.props.dispatch(rollback());
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleSweep", function () {
      _this.props.dispatch(sweep());
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleCommit", function () {
      _this.props.dispatch(commit());
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleReset", function () {
      _this.props.dispatch(reset());
    });
    return _this;
  }

  (0, _createClass2.default)(TopButtons, [{
    key: "shouldComponentUpdate",
    value: function shouldComponentUpdate(nextProps) {
      return nextProps.options !== this.props.options || nextProps.liftedState !== this.props.liftedState;
    }
  }, {
    key: "render",
    value: function render() {
      var options = this.props.options;
      var features = options.features;
      var _this$props$liftedSta = this.props.liftedState,
          computedStates = _this$props$liftedSta.computedStates,
          skippedActionIds = _this$props$liftedSta.skippedActionIds,
          isPaused = _this$props$liftedSta.isPaused,
          isLocked = _this$props$liftedSta.isLocked;
      var noStates = computedStates.length < 2;
      return /*#__PURE__*/_react.default.createElement(_ui.Toolbar, {
        borderPosition: "bottom"
      }, features.pause && /*#__PURE__*/_react.default.createElement(_RecordButton.default, {
        paused: isPaused
      }), features.persist && /*#__PURE__*/_react.default.createElement(_PersistButton.default, null), features.lock && /*#__PURE__*/_react.default.createElement(_LockButton.default, {
        locked: isLocked,
        disabled: options.lib !== 'redux'
      }), /*#__PURE__*/_react.default.createElement(_ui.Divider, null), /*#__PURE__*/_react.default.createElement(_ui.Button, {
        title: "Reset to the state you created the store with",
        tooltipPosition: "bottom",
        onClick: this.handleReset
      }, "Reset"), /*#__PURE__*/_react.default.createElement(_ui.Button, {
        title: "Roll back to the last committed state",
        tooltipPosition: "bottom",
        onClick: this.handleRollback,
        disabled: noStates
      }, "Revert"), /*#__PURE__*/_react.default.createElement(_ui.Button, {
        title: "Remove all currently disabled actions from the log",
        tooltipPosition: "bottom",
        onClick: this.handleSweep,
        disabled: skippedActionIds.length === 0
      }, "Sweep"), /*#__PURE__*/_react.default.createElement(_ui.Button, {
        title: "Remove all actions from the log,\\a and make the current state your initial state",
        tooltipPosition: "bottom",
        onClick: this.handleCommit,
        disabled: noStates
      }, "Commit"), /*#__PURE__*/_react.default.createElement(_ui.Divider, null), /*#__PURE__*/_react.default.createElement(_InstanceSelector.default, null), features.sync && /*#__PURE__*/_react.default.createElement(_SyncButton.default, null));
    }
  }]);
  return TopButtons;
}(_react.Component);

exports.default = TopButtons;
(0, _defineProperty2.default)(TopButtons, "propTypes", {
  // shouldSync: PropTypes.bool,
  liftedState: _propTypes.default.object.isRequired,
  dispatch: _propTypes.default.func.isRequired,
  options: _propTypes.default.object.isRequired
});
