// @redux-devtools/inspector-monitor/lib/cjs/ActionListRow.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _dateformat = _interopRequireDefault(require("dateformat"));

var _lodash = _interopRequireDefault(require("lodash.debounce"));

var _RightSlider = _interopRequireDefault(require("./RightSlider"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var BUTTON_SKIP = 'Skip';
var BUTTON_JUMP = 'Jump';

var ActionListRow = /*#__PURE__*/function (_PureComponent) {
  (0, _inherits2.default)(ActionListRow, _PureComponent);

  var _super = _createSuper(ActionListRow);

  function ActionListRow() {
    var _this;

    (0, _classCallCheck2.default)(this, ActionListRow);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "state", {
      hover: false
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleMouseEnter", function (e) {
      if (_this.state.hover) return;

      _this.handleMouseLeave.cancel();

      _this.handleMouseEnterDebounced(e.buttons);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleMouseEnterDebounced", (0, _lodash.default)(function (buttons) {
      if (buttons) return;

      _this.setState({
        hover: true
      });
    }, 150));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleMouseLeave", (0, _lodash.default)(function () {
      _this.handleMouseEnterDebounced.cancel();

      if (_this.state.hover) _this.setState({
        hover: false
      });
    }, 100));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleMouseDown", function (e) {
      if (e.target.className.indexOf('selectorButton') === 0) return;

      _this.handleMouseLeave();
    });
    return _this;
  }

  (0, _createClass2.default)(ActionListRow, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          styling = _this$props.styling,
          isSelected = _this$props.isSelected,
          action = _this$props.action,
          actionId = _this$props.actionId,
          isInitAction = _this$props.isInitAction,
          onSelect = _this$props.onSelect,
          timestamps = _this$props.timestamps,
          isSkipped = _this$props.isSkipped,
          isInFuture = _this$props.isInFuture,
          hideActionButtons = _this$props.hideActionButtons;
      var hover = this.state.hover;
      var timeDelta = timestamps.current - timestamps.previous;
      var showButtons = hover && !isInitAction || isSkipped;

      var isButtonSelected = function isButtonSelected(btn) {
        return btn === BUTTON_SKIP && isSkipped;
      };

      var actionType = action.type;
      if (typeof actionType === 'undefined') actionType = '<UNDEFINED>';else if (actionType === null) actionType = '<NULL>';else actionType = actionType.toString() || '<EMPTY>';
      return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
        onClick: onSelect,
        onMouseEnter: !hideActionButtons && this.handleMouseEnter,
        onMouseLeave: !hideActionButtons && this.handleMouseLeave,
        onMouseDown: this.handleMouseDown,
        onMouseUp: this.handleMouseEnter,
        "data-id": actionId
      }, styling(['actionListItem', isSelected && 'actionListItemSelected', isSkipped && 'actionListItemSkipped', isInFuture && 'actionListFromFuture'], isSelected, action)), /*#__PURE__*/_react.default.createElement("div", styling(['actionListItemName', isSkipped && 'actionListItemNameSkipped']), actionType), hideActionButtons ? /*#__PURE__*/_react.default.createElement(_RightSlider.default, {
        styling: styling,
        shown: true
      }, /*#__PURE__*/_react.default.createElement("div", styling('actionListItemTime'), timeDelta === 0 ? '+00:00:00' : (0, _dateformat.default)(timeDelta, timestamps.previous ? '+MM:ss.L' : 'h:MM:ss.L'))) : /*#__PURE__*/_react.default.createElement("div", styling('actionListItemButtons'), /*#__PURE__*/_react.default.createElement(_RightSlider.default, {
        styling: styling,
        shown: !showButtons,
        rotate: true
      }, /*#__PURE__*/_react.default.createElement("div", styling('actionListItemTime'), timeDelta === 0 ? '+00:00:00' : (0, _dateformat.default)(timeDelta, timestamps.previous ? '+MM:ss.L' : 'h:MM:ss.L'))), /*#__PURE__*/_react.default.createElement(_RightSlider.default, {
        styling: styling,
        shown: showButtons,
        rotate: true
      }, /*#__PURE__*/_react.default.createElement("div", styling('actionListItemSelector'), [BUTTON_JUMP, BUTTON_SKIP].map(function (btn) {
        return (!isInitAction || btn !== BUTTON_SKIP) && /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
          key: btn,
          onClick: function onClick(e) {
            return _this2.handleButtonClick(btn, e);
          }
        }, styling(['selectorButton', isButtonSelected(btn) && 'selectorButtonSelected', 'selectorButtonSmall'], isButtonSelected(btn), true)), btn);
      })))));
    }
  }, {
    key: "handleButtonClick",
    value: function handleButtonClick(btn, e) {
      e.stopPropagation();

      switch (btn) {
        case BUTTON_SKIP:
          this.props.onToggleClick();
          break;

        case BUTTON_JUMP:
          this.props.onJumpClick();
          break;
      }
    }
  }]);
  return ActionListRow;
}(_react.PureComponent);

exports.default = ActionListRow;
(0, _defineProperty2.default)(ActionListRow, "propTypes", {
  styling: _propTypes.default.func.isRequired,
  isSelected: _propTypes.default.bool.isRequired,
  action: _propTypes.default.object.isRequired,
  isInFuture: _propTypes.default.bool.isRequired,
  isInitAction: _propTypes.default.bool.isRequired,
  onSelect: _propTypes.default.func.isRequired,
  timestamps: _propTypes.default.shape({
    current: _propTypes.default.number.isRequired,
    previous: _propTypes.default.number.isRequired
  }).isRequired,
  isSkipped: _propTypes.default.bool.isRequired
});
