// @redux-devtools/inspector-monitor/lib/cjs/ActionPreview.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _toArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toArray"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _redux = require("./redux");

var _ActionPreviewHeader = _interopRequireDefault(require("./ActionPreviewHeader"));

var _DiffTab = _interopRequireDefault(require("./tabs/DiffTab"));

var _StateTab = _interopRequireDefault(require("./tabs/StateTab"));

var _ActionTab = _interopRequireDefault(require("./tabs/ActionTab"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var DEFAULT_TABS = [{
  name: 'Action',
  component: _ActionTab.default
}, {
  name: 'Diff',
  component: _DiffTab.default
}, {
  name: 'State',
  component: _StateTab.default
}];

var ActionPreview = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(ActionPreview, _Component);

  var _super = _createSuper(ActionPreview);

  function ActionPreview() {
    var _this;

    (0, _classCallCheck2.default)(this, ActionPreview);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "labelRenderer", function (_ref, nodeType, expanded) {
      var _ref2 = (0, _toArray2.default)(_ref),
          key = _ref2[0],
          rest = _ref2.slice(1);

      var _this$props = _this.props,
          styling = _this$props.styling,
          onInspectPath = _this$props.onInspectPath,
          inspectedPath = _this$props.inspectedPath;
      return /*#__PURE__*/_react.default.createElement("span", null, /*#__PURE__*/_react.default.createElement("span", styling('treeItemKey'), key), /*#__PURE__*/_react.default.createElement("span", (0, _extends2.default)({}, styling('treeItemPin'), {
        onClick: function onClick() {
          return onInspectPath([].concat((0, _toConsumableArray2.default)(inspectedPath.slice(0, inspectedPath.length - 1)), (0, _toConsumableArray2.default)([key].concat((0, _toConsumableArray2.default)(rest)).reverse())));
        }
      }), '(pin)'), !expanded && ': ');
    });
    return _this;
  }

  (0, _createClass2.default)(ActionPreview, [{
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          styling = _this$props2.styling,
          delta = _this$props2.delta,
          error = _this$props2.error,
          nextState = _this$props2.nextState,
          onInspectPath = _this$props2.onInspectPath,
          inspectedPath = _this$props2.inspectedPath,
          tabName = _this$props2.tabName,
          isWideLayout = _this$props2.isWideLayout,
          onSelectTab = _this$props2.onSelectTab,
          action = _this$props2.action,
          actions = _this$props2.actions,
          selectedActionId = _this$props2.selectedActionId,
          startActionId = _this$props2.startActionId,
          computedStates = _this$props2.computedStates,
          base16Theme = _this$props2.base16Theme,
          invertTheme = _this$props2.invertTheme,
          tabs = _this$props2.tabs,
          dataTypeKey = _this$props2.dataTypeKey,
          monitorState = _this$props2.monitorState,
          updateMonitorState = _this$props2.updateMonitorState;
      var renderedTabs = typeof tabs === 'function' ? tabs(DEFAULT_TABS) : tabs ? tabs : DEFAULT_TABS;

      var _ref3 = renderedTabs.find(function (tab) {
        return tab.name === tabName;
      }) || renderedTabs.find(function (tab) {
        return tab.name === _redux.DEFAULT_STATE.tabName;
      }),
          TabComponent = _ref3.component;

      return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
        key: "actionPreview"
      }, styling('actionPreview')), /*#__PURE__*/_react.default.createElement(_ActionPreviewHeader.default, {
        tabs: renderedTabs,
        styling: styling,
        inspectedPath: inspectedPath,
        onInspectPath: onInspectPath,
        tabName: tabName,
        onSelectTab: onSelectTab
      }), !error && /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
        key: "actionPreviewContent"
      }, styling('actionPreviewContent')), /*#__PURE__*/_react.default.createElement(TabComponent, {
        labelRenderer: this.labelRenderer,
        styling: styling,
        computedStates: computedStates,
        actions: actions,
        selectedActionId: selectedActionId,
        startActionId: startActionId,
        base16Theme: base16Theme,
        invertTheme: invertTheme,
        isWideLayout: isWideLayout,
        dataTypeKey: dataTypeKey,
        delta: delta,
        action: action,
        nextState: nextState,
        monitorState: monitorState,
        updateMonitorState: updateMonitorState
      })), error && /*#__PURE__*/_react.default.createElement("div", styling('stateError'), error));
    }
  }]);
  return ActionPreview;
}(_react.Component);

(0, _defineProperty2.default)(ActionPreview, "defaultProps", {
  tabName: _redux.DEFAULT_STATE.tabName
});
var _default = ActionPreview;
exports.default = _default;
