// @redux-devtools/inspector-monitor/lib/cjs/ActionListHeader.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _RightSlider = _interopRequireDefault(require("./RightSlider"));

var getActiveButtons = function getActiveButtons(hasSkippedActions) {
  return [hasSkippedActions && 'Sweep', 'Commit'].filter(function (a) {
    return !!a;
  });
};

var ActionListHeader = function ActionListHeader(_ref) {
  var styling = _ref.styling,
      onSearch = _ref.onSearch,
      hasSkippedActions = _ref.hasSkippedActions,
      hasStagedActions = _ref.hasStagedActions,
      onCommit = _ref.onCommit,
      onSweep = _ref.onSweep,
      hideMainButtons = _ref.hideMainButtons;
  return /*#__PURE__*/_react.default.createElement("div", styling('actionListHeader'), /*#__PURE__*/_react.default.createElement("input", (0, _extends2.default)({}, styling('actionListHeaderSearch'), {
    onChange: function onChange(e) {
      return onSearch(e.target.value);
    },
    placeholder: "filter..."
  })), !hideMainButtons && /*#__PURE__*/_react.default.createElement("div", styling('actionListHeaderWrapper'), /*#__PURE__*/_react.default.createElement(_RightSlider.default, {
    shown: hasStagedActions,
    styling: styling
  }, /*#__PURE__*/_react.default.createElement("div", styling('actionListHeaderSelector'), getActiveButtons(hasSkippedActions).map(function (btn) {
    return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
      key: btn,
      onClick: function onClick() {
        return {
          Commit: onCommit,
          Sweep: onSweep
        }[btn]();
      }
    }, styling(['selectorButton', 'selectorButtonSmall'], false, true)), btn);
  })))));
};

ActionListHeader.propTypes = {
  styling: _propTypes.default.func.isRequired,
  onSearch: _propTypes.default.func.isRequired,
  onCommit: _propTypes.default.func.isRequired,
  onSweep: _propTypes.default.func.isRequired,
  hideMainButtons: _propTypes.default.bool,
  hasSkippedActions: _propTypes.default.bool.isRequired,
  hasStagedActions: _propTypes.default.bool.isRequired
};
var _default = ActionListHeader;
exports.default = _default;
