// @redux-devtools/inspector-monitor/lib/cjs/createDiffPatcher.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createDiffPatcher;

var _jsondiffpatch = require("jsondiffpatch");

var defaultObjectHash = function defaultObjectHash(o, idx) {
  return o === null && '$$null' || o && (o.id || o.id === 0) && "$$id:".concat(JSON.stringify(o.id)) || o && (o._id || o._id === 0) && "$$_id:".concat(JSON.stringify(o._id)) || "$$index:".concat(idx);
};

var defaultPropertyFilter = function defaultPropertyFilter(name, context) {
  return typeof context.left[name] !== 'function' && typeof context.right[name] !== 'function';
};

var defaultDiffPatcher = new _jsondiffpatch.DiffPatcher({
  arrays: {
    detectMove: false
  },
  objectHash: defaultObjectHash,
  propertyFilter: defaultPropertyFilter
});

function createDiffPatcher(objectHash, propertyFilter) {
  if (!objectHash && !propertyFilter) {
    return defaultDiffPatcher;
  }

  return new _jsondiffpatch.DiffPatcher({
    arrays: {
      detectMove: false
    },
    objectHash: objectHash || defaultObjectHash,
    propertyFilter: propertyFilter || defaultPropertyFilter
  });
}
