// @redux-devtools/ui/lib/cjs/Button/Button.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _createStyledComponent = _interopRequireDefault(require("../utils/createStyledComponent"));

var styles = _interopRequireWildcard(require("./styles"));

var _common = require("./styles/common");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var ButtonWrapper = (0, _createStyledComponent.default)(styles, 'button');
var TooltipWrapper = (0, _createStyledComponent.default)(_common.tooltipStyle);
var CommonWrapper = (0, _createStyledComponent.default)(_common.commonStyle);

var Button = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Button, _Component);

  var _super = _createSuper(Button);

  function Button() {
    var _this;

    (0, _classCallCheck2.default)(this, Button);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onMouseUp", function (e) {
      e.currentTarget.blur();
    });
    return _this;
  }

  (0, _createClass2.default)(Button, [{
    key: "shouldComponentUpdate",
    value: function shouldComponentUpdate(nextProps) {
      return nextProps.children !== this.props.children || nextProps.disabled !== this.props.disabled || nextProps.mark !== this.props.mark || nextProps.size !== this.props.size || nextProps.primary !== this.props.primary || nextProps.tooltipPosition !== this.props.tooltipPosition || nextProps.title !== this.props.title;
    }
  }, {
    key: "render",
    value: function render() {
      var button = /*#__PURE__*/_react.default.createElement(ButtonWrapper, {
        theme: this.props.theme,
        "aria-label": this.props.title,
        primary: this.props.primary,
        disabled: this.props.disabled,
        onMouseUp: this.onMouseUp,
        onClick: this.props.onClick,
        type: this.props.type
      }, this.props.children);

      var Wrapper = this.props.title ? TooltipWrapper : CommonWrapper;
      return /*#__PURE__*/_react.default.createElement(Wrapper, {
        theme: this.props.theme,
        tooltipTitle: this.props.title,
        tooltipPosition: this.props.tooltipPosition,
        size: this.props.size,
        mark: this.props.mark
      }, button);
    }
  }]);
  return Button;
}(_react.Component);

exports.default = Button;
(0, _defineProperty2.default)(Button, "propTypes", {
  children: _propTypes.default.any.isRequired,
  title: _propTypes.default.string,
  tooltipPosition: _propTypes.default.oneOf(['top', 'bottom', 'left', 'right', 'bottom-left', 'bottom-right', 'top-left', 'top-right']),
  onClick: _propTypes.default.func,
  type: _propTypes.default.string,
  disabled: _propTypes.default.bool,
  primary: _propTypes.default.bool,
  size: _propTypes.default.oneOf(['big', 'normal', 'small']),
  mark: _propTypes.default.oneOf([false, 'base08', 'base09', 'base0A', 'base0B', 'base0C', 'base0D', 'base0E', 'base0F']),
  theme: _propTypes.default.object
});
(0, _defineProperty2.default)(Button, "defaultProps", {
  tooltipPosition: 'top'
});
