// @redux-devtools/ui/lib/cjs/ContextMenu/ContextMenu.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Default = void 0;

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _ = _interopRequireDefault(require("./"));

var _data = require("./data");

var _templateObject;

var _default = {
  title: 'ContextMenu',
  component: _.default
};
exports.default = _default;

var Container = _styledComponents.default.div(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2.default)(["\n  display: flex;\n  height: 100%;\n  width: 100%;\n  justify-content: center;\n  align-items: center;\n"])));

var Template = function Template(args) {
  return /*#__PURE__*/_react.default.createElement(Container, null, /*#__PURE__*/_react.default.createElement(_.default, args));
};

var Default = Template.bind({});
exports.Default = Default;
Default.args = {
  visible: true,
  x: 100,
  y: 100,
  items: _data.items
};
Default.argTypes = {
  visible: {
    control: {
      disable: true
    }
  },
  items: {
    control: {
      disable: true
    }
  },
  onClick: {
    control: {
      disable: true
    }
  }
};
