// @redux-devtools/ui/lib/cjs/ContextMenu/ContextMenu.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _createStyledComponent = _interopRequireDefault(require("../utils/createStyledComponent"));

var _index = _interopRequireDefault(require("./styles/index"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var ContextMenuWrapper = (0, _createStyledComponent.default)(_index.default);

function isReactButtonElement(item) {
  return item.type === 'button';
}

var ContextMenu = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(ContextMenu, _Component);

  var _super = _createSuper(ContextMenu);

  function ContextMenu() {
    var _this;

    (0, _classCallCheck2.default)(this, ContextMenu);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onMouseUp", function (e) {
      e.currentTarget.blur();
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onClick", function (e) {
      _this.props.onClick(e.currentTarget.value);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "menuRef", function (c) {
      _this.menu = c;
    });
    return _this;
  }

  (0, _createClass2.default)(ContextMenu, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.amendPosition();
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      if (prevProps.x !== this.props.x || prevProps.y !== this.props.y) {
        this.amendPosition();
      }
    }
  }, {
    key: "amendPosition",
    value: function amendPosition() {
      var _this$props = this.props,
          x = _this$props.x,
          y = _this$props.y;
      var _document$documentEle = document.documentElement,
          scrollTop = _document$documentEle.scrollTop,
          scrollLeft = _document$documentEle.scrollLeft;
      var _window = window,
          innerWidth = _window.innerWidth,
          innerHeight = _window.innerHeight;
      var rect = this.menu.getBoundingClientRect();
      var left = x + scrollLeft;
      var top = y + scrollTop;

      if (y + rect.height > innerHeight) {
        top = innerHeight - rect.height;
      }

      if (x + rect.width > innerWidth) {
        left = innerWidth - rect.width;
      }

      if (top < 0) {
        top = rect.height < innerHeight ? (innerHeight - rect.height) / 2 : 0;
      }

      if (left < 0) {
        left = rect.width < innerWidth ? (innerWidth - rect.width) / 2 : 0;
      }

      this.menu.style.top = "".concat(top, "px");
      this.menu.style.left = "".concat(left, "px");
    }
  }, {
    key: "renderItems",
    value: function renderItems() {
      var _this2 = this;

      return this.props.items.map(function (item) {
        if (isReactButtonElement(item)) return item;
        var value = item.value || item.name;
        return /*#__PURE__*/_react.default.createElement("button", {
          key: value,
          value: value,
          onMouseUp: _this2.onMouseUp,
          onClick: _this2.onClick
        }, item.name);
      });
    }
  }, {
    key: "render",
    value: function render() {
      return /*#__PURE__*/_react.default.createElement(ContextMenuWrapper, {
        ref: this.menuRef,
        left: this.props.x,
        top: this.props.y,
        visible: this.props.visible
      }, this.renderItems());
    }
  }]);
  return ContextMenu;
}(_react.Component);

exports.default = ContextMenu;
(0, _defineProperty2.default)(ContextMenu, "propTypes", {
  items: _propTypes.default.array.isRequired,
  onClick: _propTypes.default.func.isRequired,
  x: _propTypes.default.number.isRequired,
  y: _propTypes.default.number.isRequired,
  visible: _propTypes.default.bool
});
