// @redux-devtools/rtk-query-monitor/lib/cjs/containers/RtkQueryInspector.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _selectors = require("../selectors");

var _reducers = require("../reducers");

var _QueryList = require("../components/QueryList");

var _QueryForm = require("../components/QueryForm");

var _QueryPreview = require("./QueryPreview");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var RtkQueryInspector = /*#__PURE__*/function (_PureComponent) {
  (0, _inherits2.default)(RtkQueryInspector, _PureComponent);

  var _super = _createSuper(RtkQueryInspector);

  function RtkQueryInspector(props) {
    var _this;

    (0, _classCallCheck2.default)(this, RtkQueryInspector);
    _this = _super.call(this, props);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "inspectorRef", /*#__PURE__*/(0, _react.createRef)());
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "isWideIntervalRef", null);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "selectors", (0, _selectors.createInspectorSelectors)());
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "updateSizeMode", function () {
      if (_this.inspectorRef.current) {
        var isWideLayout = _this.inspectorRef.current.offsetWidth >= RtkQueryInspector.wideLayout;

        if (isWideLayout !== _this.state.isWideLayout) {
          _this.setState({
            isWideLayout: isWideLayout
          });
        }
      }
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleQueryFormValuesChange", function (values) {
      _this.props.dispatch((0, _reducers.changeQueryFormValues)(values));
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleSelectQuery", function (queryInfo) {
      _this.props.dispatch((0, _reducers.selectQueryKey)(queryInfo));
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleTabChange", function (tab) {
      _this.props.dispatch((0, _reducers.selectedPreviewTab)(tab));
    });
    _this.state = {
      isWideLayout: true,
      selectorsSource: (0, _selectors.computeSelectorSource)(props, null)
    };
    return _this;
  }

  (0, _createClass2.default)(RtkQueryInspector, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.updateSizeMode();
      this.isWideIntervalRef = setInterval(this.updateSizeMode, 300);
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      if (this.isWideIntervalRef) {
        clearTimeout(this.isWideIntervalRef);
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this$state = this.state,
          selectorsSource = _this$state.selectorsSource,
          isWideLayout = _this$state.isWideLayout;
      var styling = this.props.styleUtils.styling;
      var allVisibleRtkResourceInfos = this.selectors.selectAllVisbileQueries(selectorsSource);
      var currentResInfo = this.selectors.selectCurrentQueryInfo(selectorsSource);
      var apiStates = this.selectors.selectApiStates(selectorsSource);
      var hasNoApi = apiStates == null;
      var searchQueryRegex = this.selectors.selectSearchQueryRegex(selectorsSource);
      return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
        ref: this.inspectorRef,
        "data-wide-layout": +this.state.isWideLayout
      }, styling('inspector')), /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({}, styling('querySectionWrapper'), {
        "data-wide-layout": +this.state.isWideLayout
      }), /*#__PURE__*/_react.default.createElement(_QueryForm.QueryForm, {
        searchQueryRegex: searchQueryRegex,
        values: selectorsSource.monitorState.queryForm.values,
        onFormValuesChange: this.handleQueryFormValuesChange
      }), /*#__PURE__*/_react.default.createElement(_QueryList.QueryList, {
        onSelectQuery: this.handleSelectQuery,
        resInfos: allVisibleRtkResourceInfos,
        selectedQueryKey: selectorsSource.monitorState.selectedQueryKey
      })), /*#__PURE__*/_react.default.createElement(_QueryPreview.QueryPreview, {
        selectorsSource: this.state.selectorsSource,
        selectors: this.selectors,
        resInfo: currentResInfo,
        selectedTab: selectorsSource.monitorState.selectedPreviewTab,
        onTabChange: this.handleTabChange,
        styling: styling,
        isWideLayout: isWideLayout,
        hasNoApis: hasNoApi
      }));
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(props, state) {
      var selectorsSource = (0, _selectors.computeSelectorSource)(props, state.selectorsSource);

      if (selectorsSource !== state.selectorsSource) {
        return {
          selectorsSource: selectorsSource
        };
      }

      return null;
    }
  }]);
  return RtkQueryInspector;
}(_react.PureComponent);

(0, _defineProperty2.default)(RtkQueryInspector, "wideLayout", 600);
var _default = RtkQueryInspector;
exports.default = _default;
