// @redux-devtools/ui/lib/cjs/Container/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styledComponents = require("styled-components");

var _theme = require("../utils/theme");

var _styles = require("./styles");

var ContainerFromThemeData = function ContainerFromThemeData(_ref) {
  var themeData = _ref.themeData,
      className = _ref.className,
      children = _ref.children;
  var theme = (0, _theme.useTheme)(themeData);
  return /*#__PURE__*/_react.default.createElement(_styledComponents.ThemeProvider, {
    theme: theme
  }, /*#__PURE__*/_react.default.createElement(_styles.MainContainerWrapper, {
    className: className
  }, children));
};

var Container = function Container(_ref2) {
  var themeData = _ref2.themeData,
      className = _ref2.className,
      theme = _ref2.theme,
      children = _ref2.children;

  if (!themeData) {
    return /*#__PURE__*/_react.default.createElement(_styles.ContainerWrapper, {
      className: className,
      theme: theme
    }, children);
  }

  return /*#__PURE__*/_react.default.createElement(ContainerFromThemeData, {
    themeData: themeData,
    className: className
  }, children);
};

Container.propTypes = {
  children: _propTypes.default.node,
  themeData: _propTypes.default.any,
  theme: _propTypes.default.any,
  className: _propTypes.default.string
};
var _default = Container;
exports.default = _default;
