// @redux-devtools/app/lib/esm/containers/monitors/Dispatcher.js
import _toConsumableArray from "@babel/runtime/helpers/toConsumableArray";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject, _templateObject2, _templateObject3;

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

// Based on https://github.com/YoruNoHikage/redux-devtools-dispatch
import React, { Component } from 'react';
import styled from 'styled-components';
import { Button, Select, Editor, Toolbar } from '@redux-devtools/ui';
import { connect } from 'react-redux';
import { dispatchRemotely } from '../../actions';
export var DispatcherContainer = styled.div(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  display: flex;\n  flex-direction: column;\n  flex-shrink: 0;\n  padding-top: 2px;\n  background: ", ";\n"])), function (props) {
  return props.theme.base01;
});
export var CodeContainer = styled.div(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n  height: 75px;\n  padding-right: 6px;\n  overflow: auto;\n"])));
export var ActionContainer = styled.div(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["\n  display: table;\n  width: 100%;\n  color: ", ";\n\n  > div {\n    display: table-row;\n\n    > div:first-child {\n      width: 1px;\n      padding-left: 8px;\n      display: table-cell;\n      text-overflow: ellipsis;\n      white-space: nowrap;\n      overflow: hidden;\n    }\n\n    > div:nth-child(2) {\n      display: table-cell;\n      width: 100%;\n      padding: 6px;\n    }\n  }\n"])), function (props) {
  return props.theme.base06;
});

var Dispatcher = /*#__PURE__*/function (_Component) {
  _inherits(Dispatcher, _Component);

  var _super = _createSuper(Dispatcher);

  function Dispatcher() {
    var _this;

    _classCallCheck(this, Dispatcher);

    for (var _len = arguments.length, _args = new Array(_len), _key = 0; _key < _len; _key++) {
      _args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(_args));

    _defineProperty(_assertThisInitialized(_this), "state", {
      selected: 'default',
      customAction: _this.props.options.lib === 'redux' ? "{\n  type: ''\n}" : 'this.',
      args: [],
      rest: '[]',
      changed: false
    });

    _defineProperty(_assertThisInitialized(_this), "selectActionCreator", function (selected) {
      if (selected === 'actions-help') {
        window.open('https://github.com/reduxjs/redux-devtools/blob/main/extension/docs/' + 'API/Arguments.md#actioncreators');
        return;
      }

      var args = [];

      if (selected !== 'default') {
        args.length = _this.props.options.actionCreators[selected].args.length;
      }

      _this.setState({
        selected: selected,
        args: args,
        rest: '[]',
        changed: false
      });
    });

    _defineProperty(_assertThisInitialized(_this), "handleArg", function (argIndex) {
      return function (value) {
        var args = [].concat(_toConsumableArray(_this.state.args.slice(0, argIndex)), [value || undefined], _toConsumableArray(_this.state.args.slice(argIndex + 1)));

        _this.setState({
          args: args,
          changed: true
        });
      };
    });

    _defineProperty(_assertThisInitialized(_this), "handleRest", function (rest) {
      _this.setState({
        rest: rest,
        changed: true
      });
    });

    _defineProperty(_assertThisInitialized(_this), "handleCustomAction", function (customAction) {
      _this.setState({
        customAction: customAction,
        changed: true
      });
    });

    _defineProperty(_assertThisInitialized(_this), "dispatchAction", function () {
      var _this$state = _this.state,
          selected = _this$state.selected,
          customAction = _this$state.customAction,
          args = _this$state.args,
          rest = _this$state.rest;

      if (selected !== 'default') {
        // remove trailing `undefined` arguments
        var i = args.length - 1;

        while (i >= 0 && typeof args[i] === 'undefined') {
          args.pop();
          i--;
        }

        _this.props.dispatch({
          name: _this.props.options.actionCreators[selected].name,
          selected: selected,
          args: args,
          rest: rest
        });
      } else {
        _this.props.dispatch(customAction);
      }

      _this.setState({
        changed: false
      });
    });

    return _this;
  }

  _createClass(Dispatcher, [{
    key: "UNSAFE_componentWillReceiveProps",
    value: function UNSAFE_componentWillReceiveProps(nextProps) {
      if (this.state.selected !== 'default' && !nextProps.options.actionCreators) {
        this.setState({
          selected: 'default',
          args: []
        });
      }
    }
  }, {
    key: "shouldComponentUpdate",
    value: function shouldComponentUpdate(nextProps, nextState) {
      return nextState !== this.state || nextProps.options.actionCreators !== this.props.options.actionCreators;
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var actionCreators = this.props.options.actionCreators;
      var actionElement;

      if (this.state.selected === 'default' || !actionCreators) {
        actionElement = /*#__PURE__*/React.createElement(CodeContainer, null, /*#__PURE__*/React.createElement(Editor, {
          value: this.state.customAction,
          onChange: this.handleCustomAction
        }));
      } else {
        actionElement = /*#__PURE__*/React.createElement(ActionContainer, null, actionCreators[this.state.selected].args.map(function (param, i) {
          return /*#__PURE__*/React.createElement("div", {
            key: "".concat(param).concat(i)
          }, /*#__PURE__*/React.createElement("div", null, param), /*#__PURE__*/React.createElement(Editor, {
            lineNumbers: false,
            value: _this2.state.args[i],
            onChange: _this2.handleArg(i)
          }));
        }), /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("div", null, "...rest"), /*#__PURE__*/React.createElement(Editor, {
          lineNumbers: false,
          value: this.state.rest,
          onChange: this.handleRest
        })));
      }

      var options = [{
        value: 'default',
        label: 'Custom action'
      }];

      if (actionCreators && actionCreators.length > 0) {
        options = options.concat(actionCreators.map(function (_ref, i) {
          var name = _ref.name,
              args = _ref.args;
          return {
            value: i,
            label: "".concat(name, "(").concat(args.join(', '), ")")
          };
        }));
      } else {
        options.push({
          value: 'actions-help',
          label: 'Add your app built-in actions…'
        });
      }

      return /*#__PURE__*/React.createElement(DispatcherContainer, null, actionElement, /*#__PURE__*/React.createElement(Toolbar, null, /*#__PURE__*/React.createElement(Select, {
        menuPlacement: "top",
        onChange: function onChange(option) {
          return _this2.selectActionCreator(option.value);
        },
        value: options.find(function (option) {
          return option.value === _this2.state.selected;
        }) || options.find(function (option) {
          return option.value === 'default';
        }),
        options: options
      }), /*#__PURE__*/React.createElement(Button, {
        onClick: this.dispatchAction,
        primary: this.state.changed
      }, "Dispatch")));
    }
  }]);

  return Dispatcher;
}(Component);

var actionCreators = {
  dispatch: dispatchRemotely
};
export default connect(null, actionCreators)(Dispatcher);
