// @redux-devtools/ui/lib/esm/Tabs/Tabs.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _defineProperty from "@babel/runtime/helpers/defineProperty";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import TabsHeader from './TabsHeader';
import { TabsContainer } from './styles/common';

var Tabs = /*#__PURE__*/function (_Component) {
  _inherits(Tabs, _Component);

  var _super = _createSuper(Tabs);

  function Tabs() {
    var _this;

    _classCallCheck(this, Tabs);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "onMouseUp", function (e) {
      e.currentTarget.blur();
    });

    _defineProperty(_assertThisInitialized(_this), "onClick", function (e) {
      _this.props.onClick(e.currentTarget.value);
    });

    return _this;
  }

  _createClass(Tabs, [{
    key: "renderTabs",
    value: function renderTabs() {
      var _this2 = this;

      var tabs = this.props.tabs;
      var selected = this.props.selected;
      return tabs.map(function (tab, i) {
        var isSelected;
        var value = typeof tab.value !== 'undefined' ? tab.value : tab.name;

        if (value === selected) {
          isSelected = true;

          if (tab.component) {
            _this2.SelectedComponent = tab.component;
            if (tab.selector) _this2.selector = function () {
              return tab.selector(tab);
            };
          }
        }

        return /*#__PURE__*/React.createElement("button", {
          key: i,
          value: value,
          "data-selected": isSelected,
          onMouseUp: _this2.onMouseUp,
          onClick: _this2.onClick
        }, tab.name);
      });
    }
  }, {
    key: "render",
    value: function render() {
      var tabsHeader = /*#__PURE__*/React.createElement(TabsHeader, {
        tabs: this.renderTabs(),
        items: this.props.tabs,
        main: this.props.main,
        collapsible: this.props.collapsible,
        onClick: this.props.onClick,
        selected: this.props.selected,
        position: this.props.position
      });

      if (!this.SelectedComponent) {
        return /*#__PURE__*/React.createElement(TabsContainer, {
          position: this.props.position
        }, tabsHeader);
      }

      return /*#__PURE__*/React.createElement(TabsContainer, {
        position: this.props.position
      }, tabsHeader, /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(this.SelectedComponent, this.selector && this.selector())));
    }
  }]);

  return Tabs;
}(Component);

_defineProperty(Tabs, "propTypes", {
  tabs: PropTypes.array.isRequired,
  selected: PropTypes.string,
  main: PropTypes.bool,
  onClick: PropTypes.func.isRequired,
  collapsible: PropTypes.bool,
  position: PropTypes.oneOf(['left', 'right', 'center'])
});

_defineProperty(Tabs, "defaultProps", {
  position: 'left'
});

export { Tabs as default };
