// @redux-devtools/ui/lib/esm/Tabs/Tabs.stories.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject;

import React from 'react';
import styled from 'styled-components';
import Tabs from './';
import { tabs, simple10Tabs } from './data';
var Container = styled.div(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  display: flex;\n  height: 100%;\n  width: 100%;\n  justify-content: center;\n  align-items: center;\n"])));
export default {
  title: 'Tabs',
  component: Tabs
};

var Template = function Template(args) {
  return /*#__PURE__*/React.createElement(Container, null, /*#__PURE__*/React.createElement(Tabs, args));
};

export var Default = Template.bind({});
Default.args = {
  tabs: simple10Tabs,
  selected: '2',
  main: true,
  collapsible: true,
  position: 'left'
};
Default.argTypes = {
  tabs: {
    control: {
      disable: true
    }
  },
  onClick: {
    control: {
      disable: true
    }
  }
};
export var WithContent = Template.bind({});
WithContent.args = {
  tabs: tabs,
  selected: 'Tab2',
  main: false,
  collapsible: false,
  position: 'left'
};
WithContent.argTypes = {
  tabs: {
    control: {
      disable: true
    }
  },
  onClick: {
    control: {
      disable: true
    }
  }
};
