// @redux-devtools/rtk-query-monitor/lib/cjs/components/QueryPreviewApi.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.QueryPreviewApi = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _createStylingFromTheme = require("../styles/createStylingFromTheme");

var _TreeView = require("./TreeView");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var QueryPreviewApi = /*#__PURE__*/function (_PureComponent) {
  (0, _inherits2.default)(QueryPreviewApi, _PureComponent);

  var _super = _createSuper(QueryPreviewApi);

  function QueryPreviewApi() {
    var _this;

    (0, _classCallCheck2.default)(this, QueryPreviewApi);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "shouldExpandApiStateNode", function (keyPath, value, layer) {
      var lastKey = keyPath[keyPath.length - 1];
      return layer <= 1 && lastKey !== 'config';
    });
    return _this;
  }

  (0, _createClass2.default)(QueryPreviewApi, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          apiStats = _this$props.apiStats,
          isWideLayout = _this$props.isWideLayout,
          apiState = _this$props.apiState;

      if (!apiState) {
        return null;
      }

      var hasMutations = Object.keys(apiState.mutations).length > 0;
      var hasQueries = Object.keys(apiState.queries).length > 0;
      return /*#__PURE__*/_react.default.createElement(_createStylingFromTheme.StyleUtilsContext.Consumer, null, function (_ref) {
        var styling = _ref.styling;
        return /*#__PURE__*/_react.default.createElement("article", styling('tabContent'), /*#__PURE__*/_react.default.createElement("h2", null, apiState.config.reducerPath), /*#__PURE__*/_react.default.createElement(_TreeView.TreeView, {
          before: /*#__PURE__*/_react.default.createElement("h3", null, "State"),
          data: apiState,
          shouldExpandNode: _this2.shouldExpandApiStateNode,
          isWideLayout: isWideLayout
        }), apiStats && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_TreeView.TreeView, {
          before: /*#__PURE__*/_react.default.createElement("h3", null, "Tally"),
          data: apiStats.tally,
          isWideLayout: isWideLayout
        }), hasQueries && /*#__PURE__*/_react.default.createElement(_TreeView.TreeView, {
          before: /*#__PURE__*/_react.default.createElement("h3", null, "Queries Timings"),
          data: apiStats.timings.queries,
          isWideLayout: isWideLayout
        }), hasMutations && /*#__PURE__*/_react.default.createElement(_TreeView.TreeView, {
          before: /*#__PURE__*/_react.default.createElement("h3", null, "Mutations Timings"),
          data: apiStats.timings.mutations,
          isWideLayout: isWideLayout
        })));
      });
    }
  }]);
  return QueryPreviewApi;
}(_react.PureComponent);

exports.QueryPreviewApi = QueryPreviewApi;
