// @redux-devtools/rtk-query-monitor/lib/cjs/components/SortOrderButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SortOrderButton = SortOrderButton;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _ArrowUpIcon = require("./ArrowUpIcon");

var _createStylingFromTheme = require("../styles/createStylingFromTheme");

function SortOrderButton(_ref) {
  var isAsc = _ref.isAsc,
      onChange = _ref.onChange,
      id = _ref.id;

  var handleButtonClick = function handleButtonClick() {
    if (!isAsc) {
      onChange(true);
    } else onChange(false);
  };

  var buttonLabel = isAsc ? 'asc' : 'desc';
  var arrowStyles = {
    width: '1em',
    height: '1em',
    transform: !isAsc ? 'scaleY(-1)' : undefined
  };
  return /*#__PURE__*/_react.default.createElement(_createStylingFromTheme.StyleUtilsContext.Consumer, null, function (_ref2) {
    var styling = _ref2.styling;
    return /*#__PURE__*/_react.default.createElement("button", (0, _extends2.default)({
      type: "button",
      id: id,
      onClick: handleButtonClick,
      "aria-pressed": isAsc
    }, styling(['sortButton'])), /*#__PURE__*/_react.default.createElement(_ArrowUpIcon.ArrowUpIcon, {
      style: arrowStyles
    }), buttonLabel);
  });
}
