// @redux-devtools/ui/lib/cjs/Toolbar/Toolbar.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.WithSlider = exports.Tabs = exports.Default = void 0;

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _md = require("react-icons/md");

var _ = require("../");

var _options = require("../Select/options");

var _data = require("../Tabs/data");

var _templateObject, _templateObject2;

var Container = _styledComponents.default.div(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2.default)(["\n  display: flex;\n  height: 100%;\n  width: 100%;\n  justify-content: center;\n  align-items: center;\n"])));

var SliderContainer = _styledComponents.default.div(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteral2.default)(["\n  width: 90%;\n  height: 80px;\n"])));

var _default = {
  title: 'Toolbar',
  component: _.Toolbar
};
exports.default = _default;

var Template = function Template(_ref) {
  var borderPosition = _ref.borderPosition,
      title = _ref.title,
      tooltipPosition = _ref.tooltipPosition,
      disabled = _ref.disabled,
      onClick = _ref.onClick,
      label = _ref.label;
  return /*#__PURE__*/_react.default.createElement(Container, null, /*#__PURE__*/_react.default.createElement(_.Toolbar, {
    borderPosition: borderPosition
  }, /*#__PURE__*/_react.default.createElement(_.Button, {
    title: title,
    tooltipPosition: tooltipPosition,
    disabled: disabled,
    onClick: onClick
  }, label), /*#__PURE__*/_react.default.createElement(_.Divider, null), /*#__PURE__*/_react.default.createElement(_.Button, {
    title: title,
    tooltipPosition: tooltipPosition,
    disabled: disabled,
    onClick: onClick
  }, /*#__PURE__*/_react.default.createElement(_md.MdFiberManualRecord, null)), /*#__PURE__*/_react.default.createElement(_.Divider, null), /*#__PURE__*/_react.default.createElement(_.Spacer, null), /*#__PURE__*/_react.default.createElement(_.Select, {
    options: _options.options
  })));
};

var Default = Template.bind({});
exports.Default = Default;
Default.args = {
  borderPosition: 'top',
  title: 'Hello Tooltip',
  tooltipPosition: 'top',
  disabled: false,
  label: 'Hello Button'
};
Default.argTypes = {
  borderPosition: {
    control: {
      type: 'select',
      options: ['top', 'bottom']
    }
  },
  tooltipPosition: {
    control: {
      type: 'select',
      options: ['top', 'bottom', 'left', 'right', 'bottom-left', 'bottom-right', 'top-left', 'top-right']
    }
  },
  onClick: {
    action: 'button clicked'
  }
};

var TabsTemplate = function TabsTemplate(_ref2) {
  var title = _ref2.title,
      tooltipPosition = _ref2.tooltipPosition,
      disabled = _ref2.disabled,
      buttonOnClick = _ref2.buttonOnClick,
      label = _ref2.label,
      selected = _ref2.selected,
      main = _ref2.main,
      tabsOnClick = _ref2.tabsOnClick,
      collapsible = _ref2.collapsible,
      position = _ref2.position;
  return /*#__PURE__*/_react.default.createElement(Container, null, /*#__PURE__*/_react.default.createElement(_.Toolbar, null, /*#__PURE__*/_react.default.createElement(_.Button, {
    title: title,
    tooltipPosition: tooltipPosition,
    disabled: disabled,
    onClick: buttonOnClick
  }, label), /*#__PURE__*/_react.default.createElement(_.Tabs, {
    tabs: _data.simple10Tabs,
    selected: selected,
    main: main,
    onClick: tabsOnClick,
    collapsible: collapsible,
    position: position
  }), /*#__PURE__*/_react.default.createElement(_.Button, {
    title: title,
    tooltipPosition: tooltipPosition,
    disabled: disabled,
    onClick: buttonOnClick
  }, label)));
};

var Tabs = TabsTemplate.bind({});
exports.Tabs = Tabs;
Tabs.args = {
  title: 'Hello Tooltip',
  tooltipPosition: 'top',
  disabled: false,
  label: 'Hello Button',
  selected: '2',
  main: true,
  collapsible: true,
  position: 'center'
};
Tabs.argTypes = {
  tooltipPosition: {
    control: {
      type: 'select',
      options: ['top', 'bottom', 'left', 'right', 'bottom-left', 'bottom-right', 'top-left', 'top-right']
    }
  },
  buttonOnClick: {
    action: 'button clicked'
  },
  tabsOnClick: {
    action: 'tab selected'
  },
  position: {
    control: {
      type: 'select',
      options: ['left', 'right', 'center']
    }
  }
};

var WithSliderTemplate = function WithSliderTemplate(_ref3) {
  var title = _ref3.title,
      tooltipPosition = _ref3.tooltipPosition,
      playOnClick = _ref3.playOnClick,
      value = _ref3.value,
      min = _ref3.min,
      max = _ref3.max,
      label = _ref3.label,
      withValue = _ref3.withValue,
      onChange = _ref3.onChange,
      previousStateOnClick = _ref3.previousStateOnClick,
      nextStateOnClick = _ref3.nextStateOnClick,
      selected = _ref3.selected,
      segmentedControlOnClick = _ref3.segmentedControlOnClick;
  return /*#__PURE__*/_react.default.createElement(Container, null, /*#__PURE__*/_react.default.createElement(SliderContainer, null, /*#__PURE__*/_react.default.createElement(_.Toolbar, {
    noBorder: true,
    fullHeight: true,
    compact: true
  }, /*#__PURE__*/_react.default.createElement(_.Button, {
    title: title,
    tooltipPosition: tooltipPosition,
    onClick: playOnClick
  }, /*#__PURE__*/_react.default.createElement(_md.MdPlayArrow, null)), /*#__PURE__*/_react.default.createElement(_.Slider, {
    value: value,
    min: min,
    max: max,
    label: label,
    withValue: withValue,
    onChange: onChange
  }), /*#__PURE__*/_react.default.createElement(_.Button, {
    title: "Previous state",
    tooltipPosition: tooltipPosition,
    disabled: true,
    onClick: previousStateOnClick
  }, /*#__PURE__*/_react.default.createElement(_md.MdKeyboardArrowLeft, null)), /*#__PURE__*/_react.default.createElement(_.Button, {
    title: "Next state",
    tooltipPosition: tooltipPosition,
    onClick: nextStateOnClick
  }, /*#__PURE__*/_react.default.createElement(_md.MdKeyboardArrowRight, null)), /*#__PURE__*/_react.default.createElement(_.SegmentedControl, {
    values: ['live', '1x'],
    selected: selected,
    onClick: segmentedControlOnClick
  }))));
};

var WithSlider = WithSliderTemplate.bind({});
exports.WithSlider = WithSlider;
WithSlider.args = {
  title: 'Play',
  tooltipPosition: 'top',
  value: 80,
  min: 0,
  max: 100,
  label: 'Slider label',
  withValue: false,
  selected: 'live'
};
WithSlider.argTypes = {
  tooltipPosition: {
    control: {
      type: 'select',
      options: ['top', 'bottom', 'left', 'right', 'bottom-left', 'bottom-right', 'top-left', 'top-right']
    }
  },
  playOnClick: {
    action: 'button clicked'
  },
  onChange: {
    action: 'slider changed'
  },
  previousStateOnClick: {
    action: 'previous state clicked'
  },
  nextStateOnClick: {
    action: 'next state clicked'
  },
  selected: {
    control: {
      type: 'select',
      options: ['live', '1x']
    }
  },
  segmentedControlOnClick: {
    action: 'button selected'
  }
};
