// @redux-devtools/inspector-monitor-test-tab/lib/cjs/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getDefaultTemplates = exports.TestTab = void 0;
Object.defineProperty(exports, "reduxAvaTemplate", {
  enumerable: true,
  get: function get() {
    return _ava.default;
  }
});
Object.defineProperty(exports, "reduxJestTemplate", {
  enumerable: true,
  get: function get() {
    return _jest.default;
  }
});
Object.defineProperty(exports, "reduxMochaTemplate", {
  enumerable: true,
  get: function get() {
    return _mocha.default;
  }
});
Object.defineProperty(exports, "reduxTapeTemplate", {
  enumerable: true,
  get: function get() {
    return _tape.default;
  }
});
Object.defineProperty(exports, "vanillaAvaTemplate", {
  enumerable: true,
  get: function get() {
    return _ava2.default;
  }
});
Object.defineProperty(exports, "vanillaJestTemplate", {
  enumerable: true,
  get: function get() {
    return _jest2.default;
  }
});
Object.defineProperty(exports, "vanillaMochaTemplate", {
  enumerable: true,
  get: function get() {
    return _mocha2.default;
  }
});
Object.defineProperty(exports, "vanillaTapeTemplate", {
  enumerable: true,
  get: function get() {
    return _tape2.default;
  }
});

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _ui = require("@redux-devtools/ui");

var _md = require("react-icons/md");

var _templateForm = require("./templateForm");

var _TestGenerator = _interopRequireDefault(require("./TestGenerator"));

var _template = _interopRequireDefault(require("./redux/jest/template"));

var _template2 = _interopRequireDefault(require("./redux/mocha/template"));

var _template3 = _interopRequireDefault(require("./redux/tape/template"));

var _template4 = _interopRequireDefault(require("./redux/ava/template"));

var _ava = _interopRequireDefault(require("./redux/ava"));

var _jest = _interopRequireDefault(require("./redux/jest"));

var _mocha = _interopRequireDefault(require("./redux/mocha"));

var _tape = _interopRequireDefault(require("./redux/tape"));

var _ava2 = _interopRequireDefault(require("./vanilla/ava"));

var _jest2 = _interopRequireDefault(require("./vanilla/jest"));

var _mocha2 = _interopRequireDefault(require("./vanilla/mocha"));

var _tape2 = _interopRequireDefault(require("./vanilla/tape"));

var _excluded = ["monitorState", "updateMonitorState"];

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var getDefaultTemplates = function
  /* lib */
getDefaultTemplates() {
  return (
    /*
     if (lib === 'redux') {
     return [mochaTemplate, tapeTemplate, avaTemplate];
     }
     return [mochaVTemplate, tapeVTemplate, avaVTemplate];
     */
    [_template.default, _template2.default, _template3.default, _template4.default]
  );
};

exports.getDefaultTemplates = getDefaultTemplates;

var TestTab = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(TestTab, _Component);

  var _super = _createSuper(TestTab);

  function TestTab() {
    var _this;

    (0, _classCallCheck2.default)(this, TestTab);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "state", {
      dialogStatus: null
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "getPersistedState", function () {
      return _this.props.monitorState.testGenerator || {};
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleSelectTemplate", function (selectedTemplate) {
      var _this$getPersistedSta = _this.getPersistedState(),
          _this$getPersistedSta2 = _this$getPersistedSta.templates,
          templates = _this$getPersistedSta2 === void 0 ? getDefaultTemplates() : _this$getPersistedSta2;

      _this.updateState({
        selected: templates.indexOf(selectedTemplate)
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleCloseTip", function () {
      _this.updateState({
        hideTip: true
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleCloseDialog", function () {
      _this.setState({
        dialogStatus: null
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleSubmit", function (_ref) {
      var template = _ref.formData;

      var _this$getPersistedSta3 = _this.getPersistedState(),
          _this$getPersistedSta4 = _this$getPersistedSta3.templates,
          templates = _this$getPersistedSta4 === void 0 ? getDefaultTemplates() : _this$getPersistedSta4,
          _this$getPersistedSta5 = _this$getPersistedSta3.selected,
          selected = _this$getPersistedSta5 === void 0 ? 0 : _this$getPersistedSta5;

      if (_this.state.dialogStatus === 'Add') {
        _this.updateState({
          selected: templates.length,
          templates: [].concat((0, _toConsumableArray2.default)(templates), [template])
        });
      } else {
        var editedTemplates = (0, _toConsumableArray2.default)(templates);
        editedTemplates[selected] = template;

        _this.updateState({
          templates: editedTemplates
        });
      }

      _this.handleCloseDialog();
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleRemove", function () {
      var _this$getPersistedSta6 = _this.getPersistedState(),
          _this$getPersistedSta7 = _this$getPersistedSta6.templates,
          templates = _this$getPersistedSta7 === void 0 ? getDefaultTemplates() : _this$getPersistedSta7,
          _this$getPersistedSta8 = _this$getPersistedSta6.selected,
          selected = _this$getPersistedSta8 === void 0 ? 0 : _this$getPersistedSta8;

      _this.updateState({
        selected: 0,
        templates: templates.length === 1 ? undefined : [].concat((0, _toConsumableArray2.default)(templates.slice(0, selected)), (0, _toConsumableArray2.default)(templates.slice(selected + 1)))
      });

      _this.handleCloseDialog();
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "addTemplate", function () {
      _this.setState({
        dialogStatus: 'Add'
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "editTemplate", function () {
      _this.setState({
        dialogStatus: 'Edit'
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "updateState", function (newState) {
      _this.props.updateMonitorState({
        testGenerator: _objectSpread(_objectSpread({}, _this.props.monitorState.testGenerator), newState)
      });
    });
    return _this;
  }

  (0, _createClass2.default)(TestTab, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          monitorState = _this$props.monitorState,
          updateMonitorState = _this$props.updateMonitorState,
          rest = (0, _objectWithoutProperties2.default)(_this$props, _excluded); // eslint-disable-line no-unused-vars, max-len

      var dialogStatus = this.state.dialogStatus;
      var persistedState = this.getPersistedState();
      var _persistedState$selec = persistedState.selected,
          selected = _persistedState$selec === void 0 ? 0 : _persistedState$selec,
          _persistedState$templ = persistedState.templates,
          templates = _persistedState$templ === void 0 ? getDefaultTemplates() : _persistedState$templ;
      var template = templates[selected];
      var name = template.name,
          assertion = template.assertion,
          dispatcher = template.dispatcher,
          wrap = template.wrap;
      return /*#__PURE__*/_react.default.createElement(_ui.Container, null, /*#__PURE__*/_react.default.createElement(_ui.Toolbar, null, /*#__PURE__*/_react.default.createElement("div", {
        style: {
          flexGrow: 1
        }
      }, /*#__PURE__*/_react.default.createElement(_ui.Select, {
        options: templates,
        getOptionValue: function getOptionValue(template) {
          return template.name;
        },
        getOptionLabel: function getOptionLabel(template) {
          return template.name;
        },
        value: templates.find(function (template) {
          return template.name === name;
        }),
        onChange: this.handleSelectTemplate
      })), /*#__PURE__*/_react.default.createElement(_ui.Button, {
        onClick: this.editTemplate
      }, /*#__PURE__*/_react.default.createElement(_md.MdEdit, null)), /*#__PURE__*/_react.default.createElement(_ui.Button, {
        onClick: this.addTemplate
      }, /*#__PURE__*/_react.default.createElement(_md.MdAdd, null))), !assertion ? /*#__PURE__*/_react.default.createElement(_ui.Notification, null, "No template for tests specified.") : /*#__PURE__*/_react.default.createElement(_TestGenerator.default, (0, _extends2.default)({
        isVanilla: false,
        assertion: assertion,
        dispatcher: dispatcher,
        wrap: wrap
      }, rest)), !persistedState.hideTip && assertion && rest.startActionId === null && /*#__PURE__*/_react.default.createElement(_ui.Notification, {
        onClose: this.handleCloseTip
      }, "Hold ", /*#__PURE__*/_react.default.createElement("b", null, "SHIFT"), " key to select more actions."), dialogStatus && /*#__PURE__*/_react.default.createElement(_ui.Dialog, {
        open: true,
        title: "".concat(dialogStatus, " test template"),
        onDismiss: this.handleCloseDialog,
        onSubmit: this.handleSubmit,
        actions: [/*#__PURE__*/_react.default.createElement(_ui.Button, {
          key: "cancel",
          onClick: this.handleCloseDialog
        }, "Cancel"), /*#__PURE__*/_react.default.createElement(_ui.Button, {
          key: "remove",
          onClick: this.handleRemove
        }, "Remove")],
        submitText: dialogStatus,
        schema: _templateForm.formSchema,
        uiSchema: _templateForm.uiSchema,
        formData: dialogStatus === 'Edit' ? template : _templateForm.defaultFormData
      }));
    }
  }]);
  return TestTab;
}(_react.Component);

exports.TestTab = TestTab;
(0, _defineProperty2.default)(TestTab, "propTypes", {
  monitorState: _propTypes.default.shape({
    testGenerator: _propTypes.default.shape({
      templates: _propTypes.default.array,
      selected: _propTypes.default.number,
      hideTip: _propTypes.default.bool
    })
  }).isRequired,

  /*
  options: PropTypes.shape({
    lib: PropTypes.string
  }).isRequired,
  */
  updateMonitorState: _propTypes.default.func.isRequired
});
