// @redux-devtools/inspector-monitor-test-tab/lib/cjs/templateForm.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.uiSchema = exports.formSchema = exports.defaultFormData = void 0;
var formSchema = {
  type: 'object',
  required: ['name'],
  properties: {
    name: {
      type: 'string',
      title: 'Template name'
    },
    dispatcher: {
      type: 'string',
      title: 'Dispatcher: ({ action, prevState }) => (`<template>`)'
    },
    assertion: {
      type: 'string',
      title: 'Assertion: ({ curState }) => (`<template>`)'
    },
    wrap: {
      type: 'string',
      title: 'Wrap code: ({ name, initialState, assertions }) => (`<template>`)'
    }
  }
};
exports.formSchema = formSchema;
var uiSchema = {
  dispatcher: {
    'ui:widget': 'textarea'
  },
  assertion: {
    'ui:widget': 'textarea'
  },
  wrap: {
    'ui:widget': 'textarea'
  }
};
exports.uiSchema = uiSchema;
var defaultFormData = {
  dispatcher: 'state = reducers(${prevState}, ${action});',
  assertion: 't.deepEqual(state, ${curState});',
  wrap: "test('reducers', (t) => {\n  ${assertions}\n});"
};
exports.defaultFormData = defaultFormData;
