// @redux-devtools/rtk-query-monitor/lib/esm/components/QueryPreviewApi.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _defineProperty from "@babel/runtime/helpers/defineProperty";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

import React, { PureComponent } from 'react';
import { StyleUtilsContext } from '../styles/createStylingFromTheme';
import { TreeView } from './TreeView';
export var QueryPreviewApi = /*#__PURE__*/function (_PureComponent) {
  _inherits(QueryPreviewApi, _PureComponent);

  var _super = _createSuper(QueryPreviewApi);

  function QueryPreviewApi() {
    var _this;

    _classCallCheck(this, QueryPreviewApi);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "shouldExpandApiStateNode", function (keyPath, value, layer) {
      var lastKey = keyPath[keyPath.length - 1];
      return layer <= 1 && lastKey !== 'config';
    });

    return _this;
  }

  _createClass(QueryPreviewApi, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          apiStats = _this$props.apiStats,
          isWideLayout = _this$props.isWideLayout,
          apiState = _this$props.apiState;

      if (!apiState) {
        return null;
      }

      var hasMutations = Object.keys(apiState.mutations).length > 0;
      var hasQueries = Object.keys(apiState.queries).length > 0;
      return /*#__PURE__*/React.createElement(StyleUtilsContext.Consumer, null, function (_ref) {
        var styling = _ref.styling;
        return /*#__PURE__*/React.createElement("article", styling('tabContent'), /*#__PURE__*/React.createElement("h2", null, apiState.config.reducerPath), /*#__PURE__*/React.createElement(TreeView, {
          before: /*#__PURE__*/React.createElement("h3", null, "State"),
          data: apiState,
          shouldExpandNode: _this2.shouldExpandApiStateNode,
          isWideLayout: isWideLayout
        }), apiStats && /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(TreeView, {
          before: /*#__PURE__*/React.createElement("h3", null, "Tally"),
          data: apiStats.tally,
          isWideLayout: isWideLayout
        }), hasQueries && /*#__PURE__*/React.createElement(TreeView, {
          before: /*#__PURE__*/React.createElement("h3", null, "Queries Timings"),
          data: apiStats.timings.queries,
          isWideLayout: isWideLayout
        }), hasMutations && /*#__PURE__*/React.createElement(TreeView, {
          before: /*#__PURE__*/React.createElement("h3", null, "Mutations Timings"),
          data: apiStats.timings.mutations,
          isWideLayout: isWideLayout
        })));
      });
    }
  }]);

  return QueryPreviewApi;
}(PureComponent);
