// @redux-devtools/rtk-query-monitor/lib/esm/components/SortOrderButton.js
import _extends from "@babel/runtime/helpers/extends";
import React from 'react';
import { ArrowUpIcon } from './ArrowUpIcon';
import { StyleUtilsContext } from '../styles/createStylingFromTheme';
export function SortOrderButton(_ref) {
  var isAsc = _ref.isAsc,
      onChange = _ref.onChange,
      id = _ref.id;

  var handleButtonClick = function handleButtonClick() {
    if (!isAsc) {
      onChange(true);
    } else onChange(false);
  };

  var buttonLabel = isAsc ? 'asc' : 'desc';
  var arrowStyles = {
    width: '1em',
    height: '1em',
    transform: !isAsc ? 'scaleY(-1)' : undefined
  };
  return /*#__PURE__*/React.createElement(StyleUtilsContext.Consumer, null, function (_ref2) {
    var styling = _ref2.styling;
    return /*#__PURE__*/React.createElement("button", _extends({
      type: "button",
      id: id,
      onClick: handleButtonClick,
      "aria-pressed": isAsc
    }, styling(['sortButton'])), /*#__PURE__*/React.createElement(ArrowUpIcon, {
      style: arrowStyles
    }), buttonLabel);
  });
}
