// @redux-devtools/ui/lib/cjs/Form/Form.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Default = void 0;

var _react = _interopRequireDefault(require("react"));

var _ = _interopRequireDefault(require("./"));

var _schema = require("./schema");

var _default = {
  title: 'Form',
  component: _.default
};
exports.default = _default;

var Template = function Template(args) {
  return /*#__PURE__*/_react.default.createElement(_.default, args);
};

var Default = Template.bind({});
exports.Default = Default;
Default.args = {
  formData: _schema.formData,
  schema: _schema.schema,
  uiSchema: _schema.uiSchema,
  submitText: 'Submit'
};
Default.argTypes = {
  children: {
    control: {
      disable: true
    }
  },
  primaryButton: {
    control: {
      disable: true
    }
  },
  noSubmit: {
    control: {
      disable: true
    }
  },
  widgets: {
    control: {
      disable: true
    }
  }
};
