// @redux-devtools/ui/lib/cjs/Form/widgets.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _Select = _interopRequireDefault(require("../Select"));

var _Slider = _interopRequireDefault(require("../Slider"));

var _excluded = ["options", "onChange", "value", "onBlur", "defaultValue", "tabIndex", "onFocus", "aria-invalid"],
    _excluded2 = ["schema", "disabled", "label", "options", "formContext", "registry"];

/* eslint-disable react/prop-types */
var SelectWidget = function SelectWidget(_ref) {
  var options = _ref.options,
      _onChange = _ref.onChange,
      value = _ref.value,
      onBlur = _ref.onBlur,
      defaultValue = _ref.defaultValue,
      tabIndex = _ref.tabIndex,
      onFocus = _ref.onFocus,
      ariaInvalid = _ref['aria-invalid'],
      rest = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  return /*#__PURE__*/_react.default.createElement(_Select.default, (0, _extends2.default)({
    options: options.enumOptions,
    onChange: function onChange(option) {
      _onChange(option === null || option === void 0 ? void 0 : option.value);
    },
    value: options.enumOptions.find(function (option) {
      return option.value === value;
    })
  }, rest));
};

var RangeWidget = function RangeWidget(_ref2) {
  var schema = _ref2.schema,
      disabled = _ref2.disabled,
      label = _ref2.label,
      options = _ref2.options,
      formContext = _ref2.formContext,
      registry = _ref2.registry,
      rest = (0, _objectWithoutProperties2.default)(_ref2, _excluded2);
  return /*#__PURE__*/_react.default.createElement(_Slider.default, (0, _extends2.default)({}, rest, {
    disabled: disabled,
    min: schema.minimum,
    max: schema.maximum,
    withValue: true
  }));
};

var _default = {
  SelectWidget: SelectWidget,
  RangeWidget: RangeWidget
};
exports.default = _default;
