// @redux-devtools/ui/lib/cjs/Select/Select.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _ = _interopRequireDefault(require("./"));

var _options = require("./options");

var _excluded = ["value"];

var _templateObject;

var Container = _styledComponents.default.div(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2.default)(["\n  display: flex;\n  height: 100%;\n  width: 100%;\n  justify-content: center;\n  align-items: center;\n\n  > div {\n    width: 90%;\n  }\n"])));

var _default = {
  title: 'Select',
  component: _.default
};
exports.default = _default;

// eslint-disable-next-line react/prop-types
var Template = function Template(_ref) {
  var value = _ref.value,
      args = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  return /*#__PURE__*/_react.default.createElement(Container, null, /*#__PURE__*/_react.default.createElement(_.default, (0, _extends2.default)({
    options: _options.options,
    value: _options.options.filter(function (option) {
      return option.value === value;
    })
  }, args)));
};

var Default = Template.bind({});
exports.Default = Default;
Default.args = {
  value: 'one',
  maxMenuHeight: 300,
  isClearable: false,
  isDisabled: false,
  isLoading: false,
  isMulti: false,
  isSearchable: true,
  menuPlacement: 'bottom'
};
Default.argTypes = {
  onChange: {
    action: 'selected'
  }
};
