// @redux-devtools/inspector-monitor/lib/esm/ActionList.js
import _extends from "@babel/runtime/helpers/extends";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _defineProperty from "@babel/runtime/helpers/defineProperty";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

import React, { PureComponent } from 'react';
import dragula from 'react-dragula';
import ActionListRow from './ActionListRow';
import ActionListHeader from './ActionListHeader';

function getTimestamps(actions, actionIds, actionId) {
  var idx = actionIds.indexOf(actionId);
  var prevActionId = actionIds[idx - 1];
  return {
    current: actions[actionId].timestamp,
    previous: idx ? actions[prevActionId].timestamp : 0
  };
}

var ActionList = /*#__PURE__*/function (_PureComponent) {
  _inherits(ActionList, _PureComponent);

  var _super = _createSuper(ActionList);

  function ActionList() {
    var _this;

    _classCallCheck(this, ActionList);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "getRef", function (node) {
      _this.node = node;
    });

    return _this;
  }

  _createClass(ActionList, [{
    key: "UNSAFE_componentWillReceiveProps",
    value: function UNSAFE_componentWillReceiveProps(nextProps) {
      var node = this.node;

      if (!node) {
        this.scrollDown = true;
      } else if (this.props.lastActionId !== nextProps.lastActionId) {
        var scrollTop = node.scrollTop,
            offsetHeight = node.offsetHeight,
            scrollHeight = node.scrollHeight;
        this.scrollDown = Math.abs(scrollHeight - (scrollTop + offsetHeight)) < 50;
      } else {
        this.scrollDown = false;
      }
    }
  }, {
    key: "componentDidMount",
    value: function componentDidMount() {
      var _this2 = this;

      this.scrollDown = true;
      this.scrollToBottom();
      if (!this.props.draggableActions) return;
      var container = this.node;
      this.drake = dragula([container], {
        copy: false,
        copySortSource: false,
        mirrorContainer: container,
        accepts: function accepts(el, target, source, sibling) {
          return !sibling || !!parseInt(sibling.getAttribute('data-id'));
        },
        moves: function moves(el, source, handle) {
          return !!parseInt(el.getAttribute('data-id')) && handle.className.indexOf('selectorButton') !== 0;
        }
      }).on('drop', function (el, target, source, sibling) {
        var beforeActionId = _this2.props.actionIds.length;

        if (sibling && sibling.className.indexOf('gu-mirror') === -1) {
          beforeActionId = parseInt(sibling.getAttribute('data-id'));
        }

        var actionId = parseInt(el.getAttribute('data-id'));

        _this2.props.onReorderAction(actionId, beforeActionId);
      });
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      if (this.drake) this.drake.destroy();
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate() {
      this.scrollToBottom();
    }
  }, {
    key: "scrollToBottom",
    value: function scrollToBottom() {
      if (this.scrollDown && this.node) {
        this.node.scrollTop = this.node.scrollHeight;
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          styling = _this$props.styling,
          actions = _this$props.actions,
          actionIds = _this$props.actionIds,
          isWideLayout = _this$props.isWideLayout,
          onToggleAction = _this$props.onToggleAction,
          skippedActionIds = _this$props.skippedActionIds,
          selectedActionId = _this$props.selectedActionId,
          startActionId = _this$props.startActionId,
          _onSelect = _this$props.onSelect,
          onSearch = _this$props.onSearch,
          searchValue = _this$props.searchValue,
          currentActionId = _this$props.currentActionId,
          hideMainButtons = _this$props.hideMainButtons,
          hideActionButtons = _this$props.hideActionButtons,
          onCommit = _this$props.onCommit,
          onSweep = _this$props.onSweep,
          onJumpToState = _this$props.onJumpToState;
      var lowerSearchValue = searchValue && searchValue.toLowerCase();
      var filteredActionIds = searchValue ? actionIds.filter(function (id) {
        return actions[id].action.type.toLowerCase().indexOf(lowerSearchValue) !== -1;
      }) : actionIds;
      return /*#__PURE__*/React.createElement("div", _extends({
        key: "actionList",
        "data-testid": "actionList"
      }, styling(['actionList', isWideLayout && 'actionListWide'], isWideLayout)), /*#__PURE__*/React.createElement(ActionListHeader, {
        styling: styling,
        onSearch: onSearch,
        onCommit: onCommit,
        onSweep: onSweep,
        hideMainButtons: hideMainButtons,
        hasSkippedActions: skippedActionIds.length > 0,
        hasStagedActions: actionIds.length > 1
      }), /*#__PURE__*/React.createElement("div", _extends({
        "data-testid": "actionListRows"
      }, styling('actionListRows'), {
        ref: this.getRef
      }), filteredActionIds.map(function (actionId) {
        return /*#__PURE__*/React.createElement(ActionListRow, {
          key: actionId,
          styling: styling,
          actionId: actionId,
          isInitAction: !actionId,
          isSelected: startActionId !== null && actionId >= startActionId && actionId <= selectedActionId || actionId === selectedActionId,
          isInFuture: actionIds.indexOf(actionId) > actionIds.indexOf(currentActionId),
          onSelect: function onSelect(e) {
            return _onSelect(e, actionId);
          },
          timestamps: getTimestamps(actions, actionIds, actionId),
          action: actions[actionId].action,
          onToggleClick: function onToggleClick() {
            return onToggleAction(actionId);
          },
          onJumpClick: function onJumpClick() {
            return onJumpToState(actionId);
          },
          onCommitClick: function onCommitClick() {
            return onCommit();
          },
          hideActionButtons: hideActionButtons,
          isSkipped: skippedActionIds.indexOf(actionId) !== -1
        });
      })));
    }
  }]);

  return ActionList;
}(PureComponent);

export { ActionList as default };
