// @redux-devtools/inspector-monitor/lib/esm/DevtoolsInspector.js
import _extends from "@babel/runtime/helpers/extends";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _defineProperty from "@babel/runtime/helpers/defineProperty";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { getBase16Theme, invertTheme } from 'react-base16-styling';
import { ActionCreators } from '@redux-devtools/core';
import { createStylingFromTheme, base16Themes } from './utils/createStylingFromTheme';
import ActionList from './ActionList';
import ActionPreview from './ActionPreview';
import getInspectedState from './utils/getInspectedState';
import createDiffPatcher from './createDiffPatcher';
import { reducer, updateMonitorState } from './redux'; // eslint-disable-next-line @typescript-eslint/unbound-method

var commit = ActionCreators.commit,
    sweep = ActionCreators.sweep,
    toggleAction = ActionCreators.toggleAction,
    jumpToAction = ActionCreators.jumpToAction,
    jumpToState = ActionCreators.jumpToState,
    reorderAction = ActionCreators.reorderAction;

function getLastActionId(props) {
  return props.stagedActionIds[props.stagedActionIds.length - 1];
}

function getCurrentActionId(props, monitorState) {
  return monitorState.selectedActionId === null ? props.stagedActionIds[props.currentStateIndex] : monitorState.selectedActionId;
}

function getFromState(actionIndex, stagedActionIds, computedStates, monitorState) {
  var startActionId = monitorState.startActionId;

  if (startActionId === null) {
    return actionIndex > 0 ? computedStates[actionIndex - 1] : null;
  }

  var fromStateIdx = stagedActionIds.indexOf(startActionId - 1);
  if (fromStateIdx === -1) fromStateIdx = 0;
  return computedStates[fromStateIdx];
}

function createIntermediateState(props, monitorState) {
  var supportImmutable = props.supportImmutable,
      computedStates = props.computedStates,
      stagedActionIds = props.stagedActionIds,
      actions = props.actionsById,
      diffObjectHash = props.diffObjectHash,
      diffPropertyFilter = props.diffPropertyFilter;
  var inspectedStatePath = monitorState.inspectedStatePath,
      inspectedActionPath = monitorState.inspectedActionPath;
  var currentActionId = getCurrentActionId(props, monitorState);
  var currentAction = actions[currentActionId] && actions[currentActionId].action;
  var actionIndex = stagedActionIds.indexOf(currentActionId);
  var fromState = getFromState(actionIndex, stagedActionIds, computedStates, monitorState);
  var toState = computedStates[actionIndex];
  var error = toState && toState.error;
  var fromInspectedState = !error && fromState && getInspectedState(fromState.state, inspectedStatePath, supportImmutable);
  var toInspectedState = !error && toState && getInspectedState(toState.state, inspectedStatePath, supportImmutable);
  var delta = !error && fromState && toState && createDiffPatcher(diffObjectHash, diffPropertyFilter).diff(fromInspectedState, toInspectedState);
  return {
    delta: delta,
    nextState: toState && getInspectedState(toState.state, inspectedStatePath, false),
    action: getInspectedState(currentAction, inspectedActionPath, false),
    error: error
  };
}

function createThemeState(props) {
  var base16Theme = getBase16Theme(props.theme, base16Themes);
  var theme = props.invertTheme ? invertTheme(props.theme) : props.theme;
  var styling = createStylingFromTheme(theme);
  return {
    base16Theme: base16Theme,
    styling: styling
  };
}

var DevtoolsInspector = /*#__PURE__*/function (_PureComponent) {
  _inherits(DevtoolsInspector, _PureComponent);

  var _super = _createSuper(DevtoolsInspector);

  function DevtoolsInspector() {
    var _this;

    _classCallCheck(this, DevtoolsInspector);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "state", _objectSpread(_objectSpread({}, createIntermediateState(_this.props, _this.props.monitorState)), {}, {
      isWideLayout: false,
      themeState: createThemeState(_this.props)
    }));

    _defineProperty(_assertThisInitialized(_this), "updateMonitorState", function (monitorState) {
      _this.props.dispatch(updateMonitorState(monitorState));
    });

    _defineProperty(_assertThisInitialized(_this), "inspectorCreateRef", function (node) {
      _this.inspectorRef = node;
    });

    _defineProperty(_assertThisInitialized(_this), "handleToggleAction", function (actionId) {
      _this.props.dispatch(toggleAction(actionId));
    });

    _defineProperty(_assertThisInitialized(_this), "handleJumpToState", function (actionId) {
      if (jumpToAction) {
        _this.props.dispatch(jumpToAction(actionId));
      } else {
        // Fallback for redux-devtools-instrument < 1.5
        var _index = _this.props.stagedActionIds.indexOf(actionId);

        if (_index !== -1) _this.props.dispatch(jumpToState(_index));
      }
    });

    _defineProperty(_assertThisInitialized(_this), "handleReorderAction", function (actionId, beforeActionId) {
      if (reorderAction) _this.props.dispatch(reorderAction(actionId, beforeActionId));
    });

    _defineProperty(_assertThisInitialized(_this), "handleCommit", function () {
      _this.props.dispatch(commit());
    });

    _defineProperty(_assertThisInitialized(_this), "handleSweep", function () {
      _this.props.dispatch(sweep());
    });

    _defineProperty(_assertThisInitialized(_this), "handleSearch", function (val) {
      _this.updateMonitorState({
        searchValue: val
      });
    });

    _defineProperty(_assertThisInitialized(_this), "handleSelectAction", function (e, actionId) {
      var monitorState = _this.props.monitorState;
      var startActionId;
      var selectedActionId;

      if (e.shiftKey && monitorState.selectedActionId !== null) {
        if (monitorState.startActionId !== null) {
          if (actionId >= monitorState.startActionId) {
            startActionId = Math.min(monitorState.startActionId, monitorState.selectedActionId);
            selectedActionId = actionId;
          } else {
            selectedActionId = Math.max(monitorState.startActionId, monitorState.selectedActionId);
            startActionId = actionId;
          }
        } else {
          startActionId = Math.min(actionId, monitorState.selectedActionId);
          selectedActionId = Math.max(actionId, monitorState.selectedActionId);
        }
      } else {
        startActionId = null;

        if (actionId === monitorState.selectedActionId || monitorState.startActionId !== null) {
          selectedActionId = null;
        } else {
          selectedActionId = actionId;
        }
      }

      _this.updateMonitorState({
        startActionId: startActionId,
        selectedActionId: selectedActionId
      });
    });

    _defineProperty(_assertThisInitialized(_this), "handleInspectPath", function (pathType, path) {
      _this.updateMonitorState(_defineProperty({}, pathType, path));
    });

    _defineProperty(_assertThisInitialized(_this), "handleSelectTab", function (tabName) {
      _this.updateMonitorState({
        tabName: tabName
      });
    });

    return _this;
  }

  _createClass(DevtoolsInspector, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.updateSizeMode();
      this.updateSizeTimeout = window.setInterval(this.updateSizeMode.bind(this), 150);
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      clearTimeout(this.updateSizeTimeout);
    }
  }, {
    key: "updateSizeMode",
    value: function updateSizeMode() {
      var isWideLayout = this.inspectorRef.offsetWidth > 500;

      if (isWideLayout !== this.state.isWideLayout) {
        this.setState({
          isWideLayout: isWideLayout
        });
      }
    }
  }, {
    key: "UNSAFE_componentWillReceiveProps",
    value: function UNSAFE_componentWillReceiveProps(nextProps) {
      var nextMonitorState = nextProps.monitorState;
      var monitorState = this.props.monitorState;

      if (getCurrentActionId(this.props, monitorState) !== getCurrentActionId(nextProps, nextMonitorState) || monitorState.startActionId !== nextMonitorState.startActionId || monitorState.inspectedStatePath !== nextMonitorState.inspectedStatePath || monitorState.inspectedActionPath !== nextMonitorState.inspectedActionPath || this.props.computedStates !== nextProps.computedStates || this.props.stagedActionIds !== nextProps.stagedActionIds) {
        this.setState(createIntermediateState(nextProps, nextMonitorState));
      }

      if (this.props.theme !== nextProps.theme || this.props.invertTheme !== nextProps.invertTheme) {
        this.setState({
          themeState: createThemeState(nextProps)
        });
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          actionIds = _this$props.stagedActionIds,
          actions = _this$props.actionsById,
          computedStates = _this$props.computedStates,
          draggableActions = _this$props.draggableActions,
          tabs = _this$props.tabs,
          invertTheme = _this$props.invertTheme,
          skippedActionIds = _this$props.skippedActionIds,
          currentStateIndex = _this$props.currentStateIndex,
          monitorState = _this$props.monitorState,
          dataTypeKey = _this$props.dataTypeKey,
          hideMainButtons = _this$props.hideMainButtons,
          hideActionButtons = _this$props.hideActionButtons;
      var selectedActionId = monitorState.selectedActionId,
          startActionId = monitorState.startActionId,
          searchValue = monitorState.searchValue,
          tabName = monitorState.tabName;
      var inspectedPathType = tabName === 'Action' ? 'inspectedActionPath' : 'inspectedStatePath';
      var _this$state = this.state,
          themeState = _this$state.themeState,
          isWideLayout = _this$state.isWideLayout,
          action = _this$state.action,
          nextState = _this$state.nextState,
          delta = _this$state.delta,
          error = _this$state.error;
      var base16Theme = themeState.base16Theme,
          styling = themeState.styling;
      return /*#__PURE__*/React.createElement("div", _extends({
        key: "inspector",
        "data-testid": "inspector",
        ref: this.inspectorCreateRef
      }, styling(['inspector', isWideLayout && 'inspectorWide'], isWideLayout)), /*#__PURE__*/React.createElement(ActionList, {
        actions: actions,
        actionIds: actionIds,
        isWideLayout: isWideLayout,
        searchValue: searchValue,
        selectedActionId: selectedActionId,
        startActionId: startActionId,
        skippedActionIds: skippedActionIds,
        draggableActions: draggableActions,
        hideMainButtons: hideMainButtons,
        hideActionButtons: hideActionButtons,
        styling: styling,
        onSearch: this.handleSearch,
        onSelect: this.handleSelectAction,
        onToggleAction: this.handleToggleAction,
        onJumpToState: this.handleJumpToState,
        onCommit: this.handleCommit,
        onSweep: this.handleSweep,
        onReorderAction: this.handleReorderAction,
        currentActionId: actionIds[currentStateIndex],
        lastActionId: getLastActionId(this.props)
      }), /*#__PURE__*/React.createElement(ActionPreview, {
        base16Theme: base16Theme,
        invertTheme: invertTheme,
        isWideLayout: isWideLayout,
        tabs: tabs,
        tabName: tabName,
        delta: delta,
        error: error,
        nextState: nextState,
        computedStates: computedStates,
        action: action,
        actions: actions,
        selectedActionId: selectedActionId,
        startActionId: startActionId,
        dataTypeKey: dataTypeKey,
        monitorState: this.props.monitorState,
        updateMonitorState: this.updateMonitorState,
        styling: styling,
        onInspectPath: function onInspectPath(path) {
          return _this2.handleInspectPath(inspectedPathType, path);
        },
        inspectedPath: monitorState[inspectedPathType],
        onSelectTab: this.handleSelectTab
      }));
    }
  }]);

  return DevtoolsInspector;
}(PureComponent);

_defineProperty(DevtoolsInspector, "propTypes", {
  dispatch: PropTypes.func,
  computedStates: PropTypes.array,
  stagedActionIds: PropTypes.array,
  actionsById: PropTypes.object,
  currentStateIndex: PropTypes.number,
  monitorState: PropTypes.shape({
    initialScrollTop: PropTypes.number
  }),
  preserveScrollTop: PropTypes.bool,
  draggableActions: PropTypes.bool,
  select: PropTypes.func.isRequired,
  theme: PropTypes.oneOfType([PropTypes.object, PropTypes.string]),
  supportImmutable: PropTypes.bool,
  diffObjectHash: PropTypes.func,
  diffPropertyFilter: PropTypes.func,
  hideMainButtons: PropTypes.bool,
  hideActionButtons: PropTypes.bool,
  invertTheme: PropTypes.bool,
  skippedActionIds: PropTypes.array,
  dataTypeKey: PropTypes.any,
  tabs: PropTypes.oneOfType([PropTypes.array, PropTypes.func])
});

_defineProperty(DevtoolsInspector, "update", reducer);

_defineProperty(DevtoolsInspector, "defaultProps", {
  select: function select(state) {
    return state;
  },
  supportImmutable: false,
  draggableActions: true,
  theme: 'inspector',
  invertTheme: true
});

export default DevtoolsInspector;
