// @redux-devtools/ui/lib/esm/Form/widgets.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["options", "onChange", "value", "onBlur", "defaultValue", "tabIndex", "onFocus", "aria-invalid"],
    _excluded2 = ["schema", "disabled", "label", "options", "formContext", "registry"];
import React from 'react';
import Select from '../Select';
import Slider from '../Slider';
/* eslint-disable react/prop-types */

var SelectWidget = function SelectWidget(_ref) {
  var options = _ref.options,
      _onChange = _ref.onChange,
      value = _ref.value,
      onBlur = _ref.onBlur,
      defaultValue = _ref.defaultValue,
      tabIndex = _ref.tabIndex,
      onFocus = _ref.onFocus,
      ariaInvalid = _ref['aria-invalid'],
      rest = _objectWithoutProperties(_ref, _excluded);

  return /*#__PURE__*/React.createElement(Select, _extends({
    options: options.enumOptions,
    onChange: function onChange(option) {
      _onChange(option === null || option === void 0 ? void 0 : option.value);
    },
    value: options.enumOptions.find(function (option) {
      return option.value === value;
    })
  }, rest));
};

var RangeWidget = function RangeWidget(_ref2) {
  var schema = _ref2.schema,
      disabled = _ref2.disabled,
      label = _ref2.label,
      options = _ref2.options,
      formContext = _ref2.formContext,
      registry = _ref2.registry,
      rest = _objectWithoutProperties(_ref2, _excluded2);

  return /*#__PURE__*/React.createElement(Slider, _extends({}, rest, {
    disabled: disabled,
    min: schema.minimum,
    max: schema.maximum,
    withValue: true
  }));
};

export default {
  SelectWidget: SelectWidget,
  RangeWidget: RangeWidget
};
