// @redux-devtools/app/lib/cjs/containers/monitors/Dispatcher.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.DispatcherContainer = exports.CodeContainer = exports.ActionContainer = void 0;

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _ui = require("@redux-devtools/ui");

var _reactRedux = require("react-redux");

var _actions = require("../../actions");

var _templateObject, _templateObject2, _templateObject3;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var DispatcherContainer = _styledComponents.default.div(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2.default)(["\n  display: flex;\n  flex-direction: column;\n  flex-shrink: 0;\n  padding-top: 2px;\n  background: ", ";\n"])), function (props) {
  return props.theme.base01;
});

exports.DispatcherContainer = DispatcherContainer;

var CodeContainer = _styledComponents.default.div(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteral2.default)(["\n  height: 75px;\n  padding-right: 6px;\n  overflow: auto;\n"])));

exports.CodeContainer = CodeContainer;

var ActionContainer = _styledComponents.default.div(_templateObject3 || (_templateObject3 = (0, _taggedTemplateLiteral2.default)(["\n  display: table;\n  width: 100%;\n  color: ", ";\n\n  > div {\n    display: table-row;\n\n    > div:first-child {\n      width: 1px;\n      padding-left: 8px;\n      display: table-cell;\n      text-overflow: ellipsis;\n      white-space: nowrap;\n      overflow: hidden;\n    }\n\n    > div:nth-child(2) {\n      display: table-cell;\n      width: 100%;\n      padding: 6px;\n    }\n  }\n"])), function (props) {
  return props.theme.base06;
});

exports.ActionContainer = ActionContainer;

var Dispatcher = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Dispatcher, _Component);

  var _super = _createSuper(Dispatcher);

  function Dispatcher() {
    var _this;

    (0, _classCallCheck2.default)(this, Dispatcher);

    for (var _len = arguments.length, _args = new Array(_len), _key = 0; _key < _len; _key++) {
      _args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(_args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "state", {
      selected: 'default',
      customAction: _this.props.options.lib === 'redux' ? "{\n  type: ''\n}" : 'this.',
      args: [],
      rest: '[]',
      changed: false
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "selectActionCreator", function (selected) {
      if (selected === 'actions-help') {
        window.open('https://github.com/reduxjs/redux-devtools/blob/main/extension/docs/' + 'API/Arguments.md#actioncreators');
        return;
      }

      var args = [];

      if (selected !== 'default') {
        args.length = _this.props.options.actionCreators[selected].args.length;
      }

      _this.setState({
        selected: selected,
        args: args,
        rest: '[]',
        changed: false
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleArg", function (argIndex) {
      return function (value) {
        var args = [].concat((0, _toConsumableArray2.default)(_this.state.args.slice(0, argIndex)), [value || undefined], (0, _toConsumableArray2.default)(_this.state.args.slice(argIndex + 1)));

        _this.setState({
          args: args,
          changed: true
        });
      };
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleRest", function (rest) {
      _this.setState({
        rest: rest,
        changed: true
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleCustomAction", function (customAction) {
      _this.setState({
        customAction: customAction,
        changed: true
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "dispatchAction", function () {
      var _this$state = _this.state,
          selected = _this$state.selected,
          customAction = _this$state.customAction,
          args = _this$state.args,
          rest = _this$state.rest;

      if (selected !== 'default') {
        // remove trailing `undefined` arguments
        var i = args.length - 1;

        while (i >= 0 && typeof args[i] === 'undefined') {
          args.pop();
          i--;
        }

        _this.props.dispatch({
          name: _this.props.options.actionCreators[selected].name,
          selected: selected,
          args: args,
          rest: rest
        });
      } else {
        _this.props.dispatch(customAction);
      }

      _this.setState({
        changed: false
      });
    });
    return _this;
  }

  (0, _createClass2.default)(Dispatcher, [{
    key: "UNSAFE_componentWillReceiveProps",
    value: function UNSAFE_componentWillReceiveProps(nextProps) {
      if (this.state.selected !== 'default' && !nextProps.options.actionCreators) {
        this.setState({
          selected: 'default',
          args: []
        });
      }
    }
  }, {
    key: "shouldComponentUpdate",
    value: function shouldComponentUpdate(nextProps, nextState) {
      return nextState !== this.state || nextProps.options.actionCreators !== this.props.options.actionCreators;
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var actionCreators = this.props.options.actionCreators;
      var actionElement;

      if (this.state.selected === 'default' || !actionCreators) {
        actionElement = /*#__PURE__*/_react.default.createElement(CodeContainer, null, /*#__PURE__*/_react.default.createElement(_ui.Editor, {
          value: this.state.customAction,
          onChange: this.handleCustomAction
        }));
      } else {
        actionElement = /*#__PURE__*/_react.default.createElement(ActionContainer, null, actionCreators[this.state.selected].args.map(function (param, i) {
          return /*#__PURE__*/_react.default.createElement("div", {
            key: "".concat(param).concat(i)
          }, /*#__PURE__*/_react.default.createElement("div", null, param), /*#__PURE__*/_react.default.createElement(_ui.Editor, {
            lineNumbers: false,
            value: _this2.state.args[i],
            onChange: _this2.handleArg(i)
          }));
        }), /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("div", null, "...rest"), /*#__PURE__*/_react.default.createElement(_ui.Editor, {
          lineNumbers: false,
          value: this.state.rest,
          onChange: this.handleRest
        })));
      }

      var options = [{
        value: 'default',
        label: 'Custom action'
      }];

      if (actionCreators && actionCreators.length > 0) {
        options = options.concat(actionCreators.map(function (_ref, i) {
          var name = _ref.name,
              args = _ref.args;
          return {
            value: i,
            label: "".concat(name, "(").concat(args.join(', '), ")")
          };
        }));
      } else {
        options.push({
          value: 'actions-help',
          label: 'Add your app built-in actions…'
        });
      }

      return /*#__PURE__*/_react.default.createElement(DispatcherContainer, null, actionElement, /*#__PURE__*/_react.default.createElement(_ui.Toolbar, null, /*#__PURE__*/_react.default.createElement(_ui.Select, {
        menuPlacement: "top",
        onChange: function onChange(option) {
          return _this2.selectActionCreator(option.value);
        },
        value: options.find(function (option) {
          return option.value === _this2.state.selected;
        }) || options.find(function (option) {
          return option.value === 'default';
        }),
        options: options
      }), /*#__PURE__*/_react.default.createElement(_ui.Button, {
        onClick: this.dispatchAction,
        primary: this.state.changed
      }, "Dispatch")));
    }
  }]);
  return Dispatcher;
}(_react.Component);

var actionCreators = {
  dispatch: _actions.dispatchRemotely
};

var _default = (0, _reactRedux.connect)(null, actionCreators)(Dispatcher);

exports.default = _default;
