// @redux-devtools/app/lib/esm/components/buttons/DispatcherButton.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Button } from '@redux-devtools/ui';
import { FaTerminal } from 'react-icons/fa';
import { toggleDispatcher } from '../../actions';

var DispatcherButton = /*#__PURE__*/function (_Component) {
  _inherits(DispatcherButton, _Component);

  var _super = _createSuper(DispatcherButton);

  function DispatcherButton() {
    _classCallCheck(this, DispatcherButton);

    return _super.apply(this, arguments);
  }

  _createClass(DispatcherButton, [{
    key: "shouldComponentUpdate",
    value: function shouldComponentUpdate(nextProps) {
      return nextProps.dispatcherIsOpen !== this.props.dispatcherIsOpen;
    }
  }, {
    key: "render",
    value: function render() {
      return /*#__PURE__*/React.createElement(Button, {
        mark: this.props.dispatcherIsOpen && 'base0D',
        title: this.props.dispatcherIsOpen ? 'Hide dispatcher' : 'Show dispatcher',
        onClick: this.props.toggleDispatcher,
        tooltipPosition: "top-left"
      }, /*#__PURE__*/React.createElement(FaTerminal, null));
    }
  }]);

  return DispatcherButton;
}(Component);

var actionCreators = {
  toggleDispatcher: toggleDispatcher
};
export default connect(null, actionCreators)(DispatcherButton);
