// @redux-devtools/app/lib/esm/components/buttons/ExportButton.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Button } from '@redux-devtools/ui';
import { TiDownload } from 'react-icons/ti';
import { exportState } from '../../actions';

var ExportButton = /*#__PURE__*/function (_Component) {
  _inherits(ExportButton, _Component);

  var _super = _createSuper(ExportButton);

  function ExportButton() {
    _classCallCheck(this, ExportButton);

    return _super.apply(this, arguments);
  }

  _createClass(ExportButton, [{
    key: "shouldComponentUpdate",
    value: function shouldComponentUpdate() {
      return false;
    }
  }, {
    key: "render",
    value: function render() {
      return /*#__PURE__*/React.createElement(Button, {
        title: "Export to a file",
        onClick: this.props.exportState
      }, /*#__PURE__*/React.createElement(TiDownload, null));
    }
  }]);

  return ExportButton;
}(Component);

var actionCreators = {
  exportState: exportState
};
export default connect(null, actionCreators)(ExportButton);
