// @redux-devtools/app/lib/esm/components/buttons/PersistButton.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Button } from '@redux-devtools/ui';
import { FaThumbtack } from 'react-icons/fa';
import { togglePersist } from '../../actions';

var LockButton = /*#__PURE__*/function (_Component) {
  _inherits(LockButton, _Component);

  var _super = _createSuper(LockButton);

  function LockButton() {
    _classCallCheck(this, LockButton);

    return _super.apply(this, arguments);
  }

  _createClass(LockButton, [{
    key: "shouldComponentUpdate",
    value: function shouldComponentUpdate(nextProps) {
      return nextProps.persisted !== this.props.persisted;
    }
  }, {
    key: "render",
    value: function render() {
      return /*#__PURE__*/React.createElement(Button, {
        tooltipPosition: "bottom",
        disabled: this.props.disabled,
        mark: this.props.persisted && 'base0D',
        title: this.props.persisted ? 'Disable state persisting' : 'Persist state history',
        onClick: this.props.onClick
      }, /*#__PURE__*/React.createElement(FaThumbtack, null));
    }
  }]);

  return LockButton;
}(Component);

var mapStateToProps = function mapStateToProps(state) {
  return {
    persisted: state.instances.persisted
  };
};

var actionCreators = {
  onClick: togglePersist
};
export default connect(mapStateToProps, actionCreators)(LockButton);
