// @redux-devtools/app/lib/esm/components/buttons/RecordButton.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Button } from '@redux-devtools/ui';
import { MdFiberManualRecord } from 'react-icons/md';
import { pauseRecording as _pauseRecording } from '../../actions';

var RecordButton = /*#__PURE__*/function (_Component) {
  _inherits(RecordButton, _Component);

  var _super = _createSuper(RecordButton);

  function RecordButton() {
    _classCallCheck(this, RecordButton);

    return _super.apply(this, arguments);
  }

  _createClass(RecordButton, [{
    key: "shouldComponentUpdate",
    value: function shouldComponentUpdate(nextProps) {
      return nextProps.paused !== this.props.paused;
    }
  }, {
    key: "render",
    value: function render() {
      return /*#__PURE__*/React.createElement(Button, {
        tooltipPosition: "bottom-right",
        mark: !this.props.paused && 'base08',
        title: this.props.paused ? 'Start recording' : 'Pause recording',
        onClick: this.props.pauseRecording
      }, /*#__PURE__*/React.createElement(MdFiberManualRecord, null));
    }
  }]);

  return RecordButton;
}(Component);

function mapDispatchToProps(dispatch, ownProps) {
  return {
    pauseRecording: function pauseRecording() {
      return dispatch(_pauseRecording(!ownProps.paused));
    }
  };
}

export default connect(null, mapDispatchToProps)(RecordButton);
