// @redux-devtools/app/lib/esm/components/buttons/SliderButton.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Button } from '@redux-devtools/ui';
import { MdAvTimer } from 'react-icons/md';
import { toggleSlider } from '../../actions';

var SliderButton = /*#__PURE__*/function (_Component) {
  _inherits(SliderButton, _Component);

  var _super = _createSuper(SliderButton);

  function SliderButton() {
    _classCallCheck(this, SliderButton);

    return _super.apply(this, arguments);
  }

  _createClass(SliderButton, [{
    key: "shouldComponentUpdate",
    value: function shouldComponentUpdate(nextProps) {
      return nextProps.isOpen !== this.props.isOpen;
    }
  }, {
    key: "render",
    value: function render() {
      return /*#__PURE__*/React.createElement(Button, {
        mark: this.props.isOpen && 'base0D',
        title: this.props.isOpen ? 'Hide slider' : 'Show slider',
        tooltipPosition: "top-left",
        onClick: this.props.toggleSlider
      }, /*#__PURE__*/React.createElement(MdAvTimer, null));
    }
  }]);

  return SliderButton;
}(Component);

var actionCreators = {
  toggleSlider: toggleSlider
};
export default connect(null, actionCreators)(SliderButton);
