// @redux-devtools/app/lib/esm/components/buttons/SyncButton.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Button } from '@redux-devtools/ui';
import { TiArrowSync } from 'react-icons/ti';
import { toggleSync } from '../../actions';

var SyncButton = /*#__PURE__*/function (_Component) {
  _inherits(SyncButton, _Component);

  var _super = _createSuper(SyncButton);

  function SyncButton() {
    _classCallCheck(this, SyncButton);

    return _super.apply(this, arguments);
  }

  _createClass(SyncButton, [{
    key: "shouldComponentUpdate",
    value: function shouldComponentUpdate(nextProps) {
      return nextProps.sync !== this.props.sync;
    }
  }, {
    key: "render",
    value: function render() {
      return /*#__PURE__*/React.createElement(Button, {
        title: "Sync actions",
        tooltipPosition: "bottom-left",
        onClick: this.props.onClick,
        mark: this.props.sync && 'base0B'
      }, /*#__PURE__*/React.createElement(TiArrowSync, null));
    }
  }]);

  return SyncButton;
}(Component);

var mapStateToProps = function mapStateToProps(state) {
  return {
    sync: state.instances.sync
  };
};

var actionCreators = {
  onClick: toggleSync
};
export default connect(mapStateToProps, actionCreators)(SyncButton);
