// @redux-devtools/rtk-query-monitor/lib/cjs/reducers.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.changeQueryFormValues = void 0;
exports.reducer = reducer;
exports.selectedPreviewTab = exports.selectQueryKey = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _toolkit = require("@reduxjs/toolkit");

var _types = require("./types");

var _comparators = require("./utils/comparators");

var _filters = require("./utils/filters");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var initialState = {
  queryForm: {
    values: {
      queryComparator: _comparators.QueryComparators.fulfilledTimeStamp,
      isAscendingQueryComparatorOrder: false,
      searchValue: '',
      isRegexSearch: false,
      queryFilter: _filters.QueryFilters.queryKey
    }
  },
  selectedPreviewTab: _types.QueryPreviewTabs.queryinfo,
  selectedQueryKey: null
};
var monitorSlice = (0, _toolkit.createSlice)({
  /**
   * `@@` prefix is mandatory.
   * @see lifedAction @ `packages/redux-devtools-app/src/actions/index.ts`
   */
  name: '@@rtk-query-monitor',
  initialState: initialState,
  reducers: {
    changeQueryFormValues: function changeQueryFormValues(state, action) {
      state.queryForm.values = _objectSpread(_objectSpread({}, state.queryForm.values), action.payload);
    },
    selectQueryKey: function selectQueryKey(state, action) {
      state.selectedQueryKey = {
        queryKey: action.payload.queryKey,
        reducerPath: action.payload.reducerPath
      };
    },
    selectedPreviewTab: function selectedPreviewTab(state, action) {
      state.selectedPreviewTab = action.payload;
    }
  }
});

function reducer(props, state, action) {
  return monitorSlice.reducer(state, action);
}

var _monitorSlice$actions = monitorSlice.actions,
    selectQueryKey = _monitorSlice$actions.selectQueryKey,
    changeQueryFormValues = _monitorSlice$actions.changeQueryFormValues,
    selectedPreviewTab = _monitorSlice$actions.selectedPreviewTab;
exports.selectedPreviewTab = selectedPreviewTab;
exports.changeQueryFormValues = changeQueryFormValues;
exports.selectQueryKey = selectQueryKey;
