// @redux-devtools/rtk-query-monitor/lib/cjs/selectors.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.computeSelectorSource = computeSelectorSource;
exports.createInspectorSelectors = createInspectorSelectors;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _toolkit = require("@reduxjs/toolkit");

var _types = require("./types");

var _comparators = require("./utils/comparators");

var _filters = require("./utils/filters");

var _object = require("./utils/object");

var _regexp = require("./utils/regexp");

var _rtkQuery = require("./utils/rtk-query");

function computeSelectorSource(props) {
  var previous = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;
  var computedStates = props.computedStates,
      currentStateIndex = props.currentStateIndex,
      monitorState = props.monitorState,
      actionsById = props.actionsById;
  var userState = computedStates.length > 0 ? computedStates[currentStateIndex].state : null;

  if (!previous || previous.userState !== userState || previous.monitorState !== monitorState || previous.actionsById !== actionsById) {
    return {
      userState: userState,
      monitorState: monitorState,
      currentStateIndex: currentStateIndex,
      actionsById: actionsById
    };
  }

  return previous;
}

function createInspectorSelectors() {
  var selectQueryComparator = function selectQueryComparator(_ref) {
    var monitorState = _ref.monitorState;
    return _comparators.queryComparators[monitorState.queryForm.values.queryComparator];
  };

  var selectQueryListFilter = function selectQueryListFilter(_ref2) {
    var monitorState = _ref2.monitorState;
    return _filters.queryListFilters[monitorState.queryForm.values.queryFilter];
  };

  var selectActionsById = function selectActionsById(_ref3) {
    var actionsById = _ref3.actionsById;
    return actionsById;
  };

  var selectApiStates = (0, _toolkit.createSelector)(function (_ref4) {
    var userState = _ref4.userState;
    return userState;
  }, _rtkQuery.getApiStatesOf);
  var selectAllQueries = (0, _toolkit.createSelector)(selectApiStates, _rtkQuery.extractAllApiQueries);
  var selectAllMutations = (0, _toolkit.createSelector)(selectApiStates, _rtkQuery.extractAllApiMutations);
  var selectSearchQueryRegex = (0, _toolkit.createSelector)(function (_ref5) {
    var monitorState = _ref5.monitorState;
    return monitorState.queryForm.values.searchValue;
  }, function (_ref6) {
    var monitorState = _ref6.monitorState;
    return monitorState.queryForm.values.isRegexSearch;
  }, function (searchValue, isRegexSearch) {
    if (searchValue) {
      try {
        var regexPattern = isRegexSearch ? searchValue : (0, _regexp.escapeRegExpSpecialCharacter)(searchValue);
        return new RegExp(regexPattern, 'i');
      } catch (err) {// We notify that the search regex provided is not valid
      }
    }

    return null;
  });

  var selectComparatorOrder = function selectComparatorOrder(_ref7) {
    var monitorState = _ref7.monitorState;
    return monitorState.queryForm.values.isAscendingQueryComparatorOrder;
  };

  var selectAllVisbileQueries = (0, _toolkit.createSelector)([selectQueryComparator, selectQueryListFilter, selectAllQueries, selectAllMutations, selectComparatorOrder, selectSearchQueryRegex], function (comparator, queryListFilter, queryList, mutationsList, isAscending, searchRegex) {
    var filteredList = queryListFilter(searchRegex, queryList.concat(mutationsList));
    var computedComparator = isAscending ? comparator : (0, _rtkQuery.flipComparator)(comparator);
    return filteredList.slice().sort(computedComparator);
  });
  var selectCurrentQueryInfo = (0, _toolkit.createSelector)(selectAllQueries, selectAllMutations, function (_ref8) {
    var monitorState = _ref8.monitorState;
    return monitorState.selectedQueryKey;
  }, function (allQueries, allMutations, selectedQueryKey) {
    if (!selectedQueryKey) {
      return null;
    }

    var currentQueryInfo = allQueries.find(function (query) {
      return query.queryKey === selectedQueryKey.queryKey && selectedQueryKey.reducerPath === query.reducerPath;
    }) || null;

    if (!currentQueryInfo) {
      currentQueryInfo = allMutations.find(function (mutation) {
        return mutation.queryKey === selectedQueryKey.queryKey && selectedQueryKey.reducerPath === mutation.reducerPath;
      }) || null;
    }

    return currentQueryInfo;
  });

  var selectApiOfCurrentQuery = function selectApiOfCurrentQuery(selectorsSource) {
    var _apiStates$currentQue;

    var apiStates = selectApiStates(selectorsSource);
    var currentQueryInfo = selectCurrentQueryInfo(selectorsSource);

    if (!apiStates || !currentQueryInfo) {
      return null;
    }

    return (_apiStates$currentQue = apiStates[currentQueryInfo.reducerPath]) !== null && _apiStates$currentQue !== void 0 ? _apiStates$currentQue : null;
  };

  var selectProvidedOfCurrentQuery = function selectProvidedOfCurrentQuery(selectorsSource) {
    var _selectApiOfCurrentQu, _selectApiOfCurrentQu2;

    return (_selectApiOfCurrentQu = (_selectApiOfCurrentQu2 = selectApiOfCurrentQuery(selectorsSource)) === null || _selectApiOfCurrentQu2 === void 0 ? void 0 : _selectApiOfCurrentQu2.provided) !== null && _selectApiOfCurrentQu !== void 0 ? _selectApiOfCurrentQu : null;
  };

  var selectSubscriptionsOfCurrentQuery = (0, _toolkit.createSelector)([selectApiOfCurrentQuery, selectCurrentQueryInfo], function (apiState, queryInfo) {
    if (!queryInfo || !apiState) {
      return _object.emptyRecord;
    }

    return apiState.subscriptions[queryInfo.queryKey];
  });
  var selectCurrentQueryTags = (0, _toolkit.createSelector)([selectCurrentQueryInfo, selectProvidedOfCurrentQuery], _rtkQuery.getQueryTagsOf);
  var selectApiStatsOfCurrentQuery = (0, _toolkit.createSelector)(selectApiOfCurrentQuery, function (selectorsSource) {
    return selectorsSource.actionsById;
  }, function (selectorsSource) {
    return selectorsSource.currentStateIndex;
  }, _rtkQuery.generateApiStatsOfCurrentQuery);
  var selectActionsOfCurrentQuery = (0, _toolkit.createSelector)(selectCurrentQueryInfo, selectActionsById, _rtkQuery.getActionsOfCurrentQuery);
  var selectTabCounters = (0, _toolkit.createSelector)([selectSubscriptionsOfCurrentQuery, selectActionsOfCurrentQuery, selectCurrentQueryTags], function (subscriptions, actions, tags) {
    var _ref9;

    return _ref9 = {}, (0, _defineProperty2.default)(_ref9, _types.QueryPreviewTabs.queryTags, tags.length), (0, _defineProperty2.default)(_ref9, _types.QueryPreviewTabs.querySubscriptions, Object.keys(subscriptions !== null && subscriptions !== void 0 ? subscriptions : {}).length), (0, _defineProperty2.default)(_ref9, _types.QueryPreviewTabs.apiConfig, 0), (0, _defineProperty2.default)(_ref9, _types.QueryPreviewTabs.queryinfo, 0), (0, _defineProperty2.default)(_ref9, _types.QueryPreviewTabs.actions, actions.length), _ref9;
  });
  return {
    selectQueryComparator: selectQueryComparator,
    selectApiStates: selectApiStates,
    selectAllQueries: selectAllQueries,
    selectAllVisbileQueries: selectAllVisbileQueries,
    selectSearchQueryRegex: selectSearchQueryRegex,
    selectCurrentQueryInfo: selectCurrentQueryInfo,
    selectCurrentQueryTags: selectCurrentQueryTags,
    selectApiStatsOfCurrentQuery: selectApiStatsOfCurrentQuery,
    selectSubscriptionsOfCurrentQuery: selectSubscriptionsOfCurrentQuery,
    selectApiOfCurrentQuery: selectApiOfCurrentQuery,
    selectTabCounters: selectTabCounters,
    selectActionsOfCurrentQuery: selectActionsOfCurrentQuery
  };
}
