// @redux-devtools/inspector-monitor/lib/esm/tabs/ActionTab.js
import React from 'react';
import PropTypes from 'prop-types';
import { JSONTree } from 'react-json-tree';
import _getItemString from './getItemString';
import getJsonTreeTheme from './getJsonTreeTheme';

var ActionTab = function ActionTab(_ref) {
  var action = _ref.action,
      styling = _ref.styling,
      base16Theme = _ref.base16Theme,
      invertTheme = _ref.invertTheme,
      labelRenderer = _ref.labelRenderer,
      dataTypeKey = _ref.dataTypeKey,
      isWideLayout = _ref.isWideLayout;
  return /*#__PURE__*/React.createElement(JSONTree, {
    labelRenderer: labelRenderer,
    theme: getJsonTreeTheme(base16Theme),
    data: action,
    getItemString: function getItemString(type, data) {
      return _getItemString(styling, type, data, dataTypeKey, isWideLayout);
    },
    invertTheme: invertTheme,
    hideRoot: true
  });
};

ActionTab.propTypes = {
  action: PropTypes.any.isRequired,
  styling: PropTypes.func.isRequired,
  base16Theme: PropTypes.any.isRequired,
  invertTheme: PropTypes.bool.isRequired,
  labelRenderer: PropTypes.func.isRequired,
  dataTypeKey: PropTypes.string,
  isWideLayout: PropTypes.bool.isRequired
};
export default ActionTab;
