// @redux-devtools/inspector-monitor/lib/esm/tabs/DiffTab.js
import React from 'react';
import PropTypes from 'prop-types';
import JSONDiff from './JSONDiff';

var DiffTab = function DiffTab(_ref) {
  var delta = _ref.delta,
      styling = _ref.styling,
      base16Theme = _ref.base16Theme,
      invertTheme = _ref.invertTheme,
      labelRenderer = _ref.labelRenderer,
      isWideLayout = _ref.isWideLayout,
      dataTypeKey = _ref.dataTypeKey;
  return /*#__PURE__*/React.createElement(JSONDiff, {
    delta: delta,
    styling: styling,
    base16Theme: base16Theme,
    invertTheme: invertTheme,
    labelRenderer: labelRenderer,
    isWideLayout: isWideLayout,
    dataTypeKey: dataTypeKey
  });
};

DiffTab.propTypes = {
  delta: PropTypes.any,
  styling: PropTypes.func.isRequired,
  base16Theme: PropTypes.any.isRequired,
  invertTheme: PropTypes.bool.isRequired,
  labelRenderer: PropTypes.func.isRequired,
  isWideLayout: PropTypes.bool.isRequired,
  dataTypeKey: PropTypes.oneOfType([PropTypes.string, PropTypes.symbol])
};
export default DiffTab;
